### Eclipse Workspace Patch 1.0
#P org.eclipse.amalgam.releng.build.model
Index: src/org/eclipse/amalgam/releng/build/impl/ContributionImpl.java
===================================================================
RCS file: /cvsroot/modeling/org.eclipse.amalgam/releng/org.eclipse.amalgam.releng.build.model/src/org/eclipse/amalgam/releng/build/impl/ContributionImpl.java,v
retrieving revision 1.1
diff -u -r1.1 ContributionImpl.java
--- src/org/eclipse/amalgam/releng/build/impl/ContributionImpl.java	24 Nov 2008 20:36:44 -0000	1.1
+++ src/org/eclipse/amalgam/releng/build/impl/ContributionImpl.java	20 May 2009 06:08:58 -0000
@@ -20,6 +20,7 @@
 import org.eclipse.amalgam.releng.build.Contact;
 import org.eclipse.amalgam.releng.build.Contribution;
 import org.eclipse.amalgam.releng.build.Feature;
+import org.eclipse.amalgam.releng.build.Product;
 import org.eclipse.amalgam.releng.build.Repository;
 
 import org.eclipse.emf.common.notify.Notification;
@@ -48,6 +49,7 @@
  *   <li>{@link org.eclipse.amalgam.releng.build.impl.ContributionImpl#getLabel <em>Label</em>}</li>
  *   <li>{@link org.eclipse.amalgam.releng.build.impl.ContributionImpl#getRepositories <em>Repositories</em>}</li>
  *   <li>{@link org.eclipse.amalgam.releng.build.impl.ContributionImpl#getBundles <em>Bundles</em>}</li>
+ *   <li>{@link org.eclipse.amalgam.releng.build.impl.ContributionImpl#getProducts <em>Products</em>}</li>
  * </ul>
  * </p>
  *
@@ -115,6 +117,16 @@
 	protected EList<Bundle> bundles;
 
 	/**
+	 * The cached value of the '{@link #getProducts() <em>Products</em>}' containment reference list.
+	 * <!-- begin-user-doc -->
+	 * <!-- end-user-doc -->
+	 * @see #getProducts()
+	 * @generated
+	 * @ordered
+	 */
+	protected EList<Product> products;
+
+	/**
 	 * <!-- begin-user-doc -->
 	 * <!-- end-user-doc -->
 	 * @generated
@@ -207,6 +219,18 @@
 	 * <!-- end-user-doc -->
 	 * @generated
 	 */
+	public EList<Product> getProducts() {
+		if (products == null) {
+			products = new EObjectContainmentEList.Resolving<Product>(Product.class, this, BuildPackage.CONTRIBUTION__PRODUCTS);
+		}
+		return products;
+	}
+
+	/**
+	 * <!-- begin-user-doc -->
+	 * <!-- end-user-doc -->
+	 * @generated
+	 */
 	@Override
 	public NotificationChain eInverseRemove(InternalEObject otherEnd, int featureID, NotificationChain msgs) {
 		switch (featureID) {
@@ -218,6 +242,8 @@
 				return ((InternalEList<?>)getRepositories()).basicRemove(otherEnd, msgs);
 			case BuildPackage.CONTRIBUTION__BUNDLES:
 				return ((InternalEList<?>)getBundles()).basicRemove(otherEnd, msgs);
+			case BuildPackage.CONTRIBUTION__PRODUCTS:
+				return ((InternalEList<?>)getProducts()).basicRemove(otherEnd, msgs);
 		}
 		return super.eInverseRemove(otherEnd, featureID, msgs);
 	}
@@ -240,6 +266,8 @@
 				return getRepositories();
 			case BuildPackage.CONTRIBUTION__BUNDLES:
 				return getBundles();
+			case BuildPackage.CONTRIBUTION__PRODUCTS:
+				return getProducts();
 		}
 		return super.eGet(featureID, resolve, coreType);
 	}
@@ -272,6 +300,10 @@
 				getBundles().clear();
 				getBundles().addAll((Collection<? extends Bundle>)newValue);
 				return;
+			case BuildPackage.CONTRIBUTION__PRODUCTS:
+				getProducts().clear();
+				getProducts().addAll((Collection<? extends Product>)newValue);
+				return;
 		}
 		super.eSet(featureID, newValue);
 	}
@@ -299,6 +331,9 @@
 			case BuildPackage.CONTRIBUTION__BUNDLES:
 				getBundles().clear();
 				return;
+			case BuildPackage.CONTRIBUTION__PRODUCTS:
+				getProducts().clear();
+				return;
 		}
 		super.eUnset(featureID);
 	}
@@ -321,6 +356,8 @@
 				return repositories != null && !repositories.isEmpty();
 			case BuildPackage.CONTRIBUTION__BUNDLES:
 				return bundles != null && !bundles.isEmpty();
+			case BuildPackage.CONTRIBUTION__PRODUCTS:
+				return products != null && !products.isEmpty();
 		}
 		return super.eIsSet(featureID);
 	}
Index: src/org/eclipse/amalgam/releng/build/impl/BuildFactoryImpl.java
===================================================================
RCS file: /cvsroot/modeling/org.eclipse.amalgam/releng/org.eclipse.amalgam.releng.build.model/src/org/eclipse/amalgam/releng/build/impl/BuildFactoryImpl.java,v
retrieving revision 1.2
diff -u -r1.2 BuildFactoryImpl.java
--- src/org/eclipse/amalgam/releng/build/impl/BuildFactoryImpl.java	6 Dec 2008 03:59:50 -0000	1.2
+++ src/org/eclipse/amalgam/releng/build/impl/BuildFactoryImpl.java	20 May 2009 06:08:58 -0000
@@ -28,6 +28,7 @@
 import org.eclipse.amalgam.releng.build.Map;
 import org.eclipse.amalgam.releng.build.OS;
 import org.eclipse.amalgam.releng.build.Platform;
+import org.eclipse.amalgam.releng.build.Product;
 import org.eclipse.amalgam.releng.build.Promotion;
 import org.eclipse.amalgam.releng.build.Repository;
 import org.eclipse.amalgam.releng.build.WS;
@@ -96,6 +97,7 @@
 			case BuildPackage.BUNDLE: return createBundle();
 			case BuildPackage.COMPILER: return createCompiler();
 			case BuildPackage.PROMOTION: return createPromotion();
+			case BuildPackage.PRODUCT: return createProduct();
 			default:
 				throw new IllegalArgumentException("The class '" + eClass.getName() + "' is not a valid classifier");
 		}
@@ -272,6 +274,16 @@
 	 * <!-- end-user-doc -->
 	 * @generated
 	 */
+	public Product createProduct() {
+		ProductImpl product = new ProductImpl();
+		return product;
+	}
+
+	/**
+	 * <!-- begin-user-doc -->
+	 * <!-- end-user-doc -->
+	 * @generated
+	 */
 	public BuildType createBuildTypeFromString(EDataType eDataType, String initialValue) {
 		BuildType result = BuildType.get(initialValue);
 		if (result == null) throw new IllegalArgumentException("The value '" + initialValue + "' is not a valid enumerator of '" + eDataType.getName() + "'");
Index: src/org/eclipse/amalgam/releng/build/impl/FeatureImpl.java
===================================================================
RCS file: /cvsroot/modeling/org.eclipse.amalgam/releng/org.eclipse.amalgam.releng.build.model/src/org/eclipse/amalgam/releng/build/impl/FeatureImpl.java,v
retrieving revision 1.3
diff -u -r1.3 FeatureImpl.java
--- src/org/eclipse/amalgam/releng/build/impl/FeatureImpl.java	17 Apr 2009 23:42:56 -0000	1.3
+++ src/org/eclipse/amalgam/releng/build/impl/FeatureImpl.java	20 May 2009 06:08:58 -0000
@@ -38,57 +38,14 @@
  * <p>
  * The following features are implemented:
  * <ul>
- *   <li>{@link org.eclipse.amalgam.releng.build.impl.FeatureImpl#getId <em>Id</em>}</li>
- *   <li>{@link org.eclipse.amalgam.releng.build.impl.FeatureImpl#getVersion <em>Version</em>}</li>
  *   <li>{@link org.eclipse.amalgam.releng.build.impl.FeatureImpl#getCategory <em>Category</em>}</li>
- *   <li>{@link org.eclipse.amalgam.releng.build.impl.FeatureImpl#getRepo <em>Repo</em>}</li>
  *   <li>{@link org.eclipse.amalgam.releng.build.impl.FeatureImpl#isInProduct <em>In Product</em>}</li>
  * </ul>
  * </p>
  *
  * @generated
  */
-public class FeatureImpl extends EObjectImpl implements Feature {
-	/**
-	 * The default value of the '{@link #getId() <em>Id</em>}' attribute.
-	 * <!-- begin-user-doc -->
-	 * <!-- end-user-doc -->
-	 * @see #getId()
-	 * @generated
-	 * @ordered
-	 */
-	protected static final String ID_EDEFAULT = null;
-
-	/**
-	 * The cached value of the '{@link #getId() <em>Id</em>}' attribute.
-	 * <!-- begin-user-doc -->
-	 * <!-- end-user-doc -->
-	 * @see #getId()
-	 * @generated
-	 * @ordered
-	 */
-	protected String id = ID_EDEFAULT;
-
-	/**
-	 * The default value of the '{@link #getVersion() <em>Version</em>}' attribute.
-	 * <!-- begin-user-doc -->
-	 * <!-- end-user-doc -->
-	 * @see #getVersion()
-	 * @generated
-	 * @ordered
-	 */
-	protected static final String VERSION_EDEFAULT = null;
-
-	/**
-	 * The cached value of the '{@link #getVersion() <em>Version</em>}' attribute.
-	 * <!-- begin-user-doc -->
-	 * <!-- end-user-doc -->
-	 * @see #getVersion()
-	 * @generated
-	 * @ordered
-	 */
-	protected String version = VERSION_EDEFAULT;
-
+public class FeatureImpl extends InstallationUnitImpl implements Feature {
 	/**
 	 * The cached value of the '{@link #getCategory() <em>Category</em>}' reference list.
 	 * <!-- begin-user-doc -->
@@ -100,16 +57,6 @@
 	protected EList<Category> category;
 
 	/**
-	 * The cached value of the '{@link #getRepo() <em>Repo</em>}' reference.
-	 * <!-- begin-user-doc -->
-	 * <!-- end-user-doc -->
-	 * @see #getRepo()
-	 * @generated
-	 * @ordered
-	 */
-	protected Repository repo;
-
-	/**
 	 * The default value of the '{@link #isInProduct() <em>In Product</em>}' attribute.
 	 * <!-- begin-user-doc -->
 	 * <!-- end-user-doc -->
@@ -153,48 +100,6 @@
 	 * <!-- end-user-doc -->
 	 * @generated
 	 */
-	public String getId() {
-		return id;
-	}
-
-	/**
-	 * <!-- begin-user-doc -->
-	 * <!-- end-user-doc -->
-	 * @generated
-	 */
-	public void setId(String newId) {
-		String oldId = id;
-		id = newId;
-		if (eNotificationRequired())
-			eNotify(new ENotificationImpl(this, Notification.SET, BuildPackage.FEATURE__ID, oldId, id));
-	}
-
-	/**
-	 * <!-- begin-user-doc -->
-	 * <!-- end-user-doc -->
-	 * @generated
-	 */
-	public String getVersion() {
-		return version;
-	}
-
-	/**
-	 * <!-- begin-user-doc -->
-	 * <!-- end-user-doc -->
-	 * @generated
-	 */
-	public void setVersion(String newVersion) {
-		String oldVersion = version;
-		version = newVersion;
-		if (eNotificationRequired())
-			eNotify(new ENotificationImpl(this, Notification.SET, BuildPackage.FEATURE__VERSION, oldVersion, version));
-	}
-
-	/**
-	 * <!-- begin-user-doc -->
-	 * <!-- end-user-doc -->
-	 * @generated
-	 */
 	public EList<Category> getCategory() {
 		if (category == null) {
 			category = new EObjectWithInverseResolvingEList.ManyInverse<Category>(Category.class, this, BuildPackage.FEATURE__CATEGORY, BuildPackage.CATEGORY__FEATURES);
@@ -207,44 +112,6 @@
 	 * <!-- end-user-doc -->
 	 * @generated
 	 */
-	public Repository getRepo() {
-		if (repo != null && repo.eIsProxy()) {
-			InternalEObject oldRepo = (InternalEObject)repo;
-			repo = (Repository)eResolveProxy(oldRepo);
-			if (repo != oldRepo) {
-				if (eNotificationRequired())
-					eNotify(new ENotificationImpl(this, Notification.RESOLVE, BuildPackage.FEATURE__REPO, oldRepo, repo));
-			}
-		}
-		return repo;
-	}
-
-	/**
-	 * <!-- begin-user-doc -->
-	 * <!-- end-user-doc -->
-	 * @generated
-	 */
-	public Repository basicGetRepo() {
-		return repo;
-	}
-
-	/**
-	 * <!-- begin-user-doc -->
-	 * <!-- end-user-doc -->
-	 * @generated
-	 */
-	public void setRepo(Repository newRepo) {
-		Repository oldRepo = repo;
-		repo = newRepo;
-		if (eNotificationRequired())
-			eNotify(new ENotificationImpl(this, Notification.SET, BuildPackage.FEATURE__REPO, oldRepo, repo));
-	}
-
-	/**
-	 * <!-- begin-user-doc -->
-	 * <!-- end-user-doc -->
-	 * @generated
-	 */
 	public boolean isInProduct() {
 		return inProduct;
 	}
@@ -298,15 +165,8 @@
 	@Override
 	public Object eGet(int featureID, boolean resolve, boolean coreType) {
 		switch (featureID) {
-			case BuildPackage.FEATURE__ID:
-				return getId();
-			case BuildPackage.FEATURE__VERSION:
-				return getVersion();
 			case BuildPackage.FEATURE__CATEGORY:
 				return getCategory();
-			case BuildPackage.FEATURE__REPO:
-				if (resolve) return getRepo();
-				return basicGetRepo();
 			case BuildPackage.FEATURE__IN_PRODUCT:
 				return isInProduct();
 		}
@@ -322,19 +182,10 @@
 	@Override
 	public void eSet(int featureID, Object newValue) {
 		switch (featureID) {
-			case BuildPackage.FEATURE__ID:
-				setId((String)newValue);
-				return;
-			case BuildPackage.FEATURE__VERSION:
-				setVersion((String)newValue);
-				return;
 			case BuildPackage.FEATURE__CATEGORY:
 				getCategory().clear();
 				getCategory().addAll((Collection<? extends Category>)newValue);
 				return;
-			case BuildPackage.FEATURE__REPO:
-				setRepo((Repository)newValue);
-				return;
 			case BuildPackage.FEATURE__IN_PRODUCT:
 				setInProduct((Boolean)newValue);
 				return;
@@ -350,18 +201,9 @@
 	@Override
 	public void eUnset(int featureID) {
 		switch (featureID) {
-			case BuildPackage.FEATURE__ID:
-				setId(ID_EDEFAULT);
-				return;
-			case BuildPackage.FEATURE__VERSION:
-				setVersion(VERSION_EDEFAULT);
-				return;
 			case BuildPackage.FEATURE__CATEGORY:
 				getCategory().clear();
 				return;
-			case BuildPackage.FEATURE__REPO:
-				setRepo((Repository)null);
-				return;
 			case BuildPackage.FEATURE__IN_PRODUCT:
 				setInProduct(IN_PRODUCT_EDEFAULT);
 				return;
@@ -377,14 +219,8 @@
 	@Override
 	public boolean eIsSet(int featureID) {
 		switch (featureID) {
-			case BuildPackage.FEATURE__ID:
-				return ID_EDEFAULT == null ? id != null : !ID_EDEFAULT.equals(id);
-			case BuildPackage.FEATURE__VERSION:
-				return VERSION_EDEFAULT == null ? version != null : !VERSION_EDEFAULT.equals(version);
 			case BuildPackage.FEATURE__CATEGORY:
 				return category != null && !category.isEmpty();
-			case BuildPackage.FEATURE__REPO:
-				return repo != null;
 			case BuildPackage.FEATURE__IN_PRODUCT:
 				return inProduct != IN_PRODUCT_EDEFAULT;
 		}
@@ -401,11 +237,7 @@
 		if (eIsProxy()) return super.toString();
 
 		StringBuffer result = new StringBuffer(super.toString());
-		result.append(" (id: ");
-		result.append(id);
-		result.append(", version: ");
-		result.append(version);
-		result.append(", inProduct: ");
+		result.append(" (inProduct: ");
 		result.append(inProduct);
 		result.append(')');
 		return result.toString();
Index: src/org/eclipse/amalgam/releng/build/impl/BuildPackageImpl.java
===================================================================
RCS file: /cvsroot/modeling/org.eclipse.amalgam/releng/org.eclipse.amalgam.releng.build.model/src/org/eclipse/amalgam/releng/build/impl/BuildPackageImpl.java,v
retrieving revision 1.10
diff -u -r1.10 BuildPackageImpl.java
--- src/org/eclipse/amalgam/releng/build/impl/BuildPackageImpl.java	17 Apr 2009 23:42:56 -0000	1.10
+++ src/org/eclipse/amalgam/releng/build/impl/BuildPackageImpl.java	20 May 2009 06:08:58 -0000
@@ -24,8 +24,10 @@
 import org.eclipse.amalgam.releng.build.Contact;
 import org.eclipse.amalgam.releng.build.Contribution;
 import org.eclipse.amalgam.releng.build.Feature;
+import org.eclipse.amalgam.releng.build.InstallationUnit;
 import org.eclipse.amalgam.releng.build.Map;
 import org.eclipse.amalgam.releng.build.Platform;
+import org.eclipse.amalgam.releng.build.Product;
 import org.eclipse.amalgam.releng.build.Promotion;
 import org.eclipse.amalgam.releng.build.Repository;
 
@@ -137,6 +139,20 @@
 	 * <!-- end-user-doc -->
 	 * @generated
 	 */
+	private EClass installationUnitEClass = null;
+
+	/**
+	 * <!-- begin-user-doc -->
+	 * <!-- end-user-doc -->
+	 * @generated
+	 */
+	private EClass productEClass = null;
+
+	/**
+	 * <!-- begin-user-doc -->
+	 * <!-- end-user-doc -->
+	 * @generated
+	 */
 	private EEnum buildTypeEEnum = null;
 
 	/**
@@ -219,7 +235,7 @@
 		if (isInited) return (BuildPackage)EPackage.Registry.INSTANCE.getEPackage(BuildPackage.eNS_URI);
 
 		// Obtain or create and register package
-		BuildPackageImpl theBuildPackage = (BuildPackageImpl)(EPackage.Registry.INSTANCE.getEPackage(eNS_URI) instanceof BuildPackageImpl ? EPackage.Registry.INSTANCE.getEPackage(eNS_URI) : new BuildPackageImpl());
+		BuildPackageImpl theBuildPackage = (BuildPackageImpl)(EPackage.Registry.INSTANCE.get(eNS_URI) instanceof BuildPackageImpl ? EPackage.Registry.INSTANCE.get(eNS_URI) : new BuildPackageImpl());
 
 		isInited = true;
 
@@ -694,6 +710,15 @@
 	 * <!-- end-user-doc -->
 	 * @generated
 	 */
+	public EReference getContribution_Products() {
+		return (EReference)contributionEClass.getEStructuralFeatures().get(5);
+	}
+
+	/**
+	 * <!-- begin-user-doc -->
+	 * <!-- end-user-doc -->
+	 * @generated
+	 */
 	public EClass getContact() {
 		return contactEClass;
 	}
@@ -730,8 +755,8 @@
 	 * <!-- end-user-doc -->
 	 * @generated
 	 */
-	public EAttribute getFeature_Id() {
-		return (EAttribute)featureEClass.getEStructuralFeatures().get(0);
+	public EReference getFeature_Category() {
+		return (EReference)featureEClass.getEStructuralFeatures().get(0);
 	}
 
 	/**
@@ -739,7 +764,7 @@
 	 * <!-- end-user-doc -->
 	 * @generated
 	 */
-	public EAttribute getFeature_Version() {
+	public EAttribute getFeature_InProduct() {
 		return (EAttribute)featureEClass.getEStructuralFeatures().get(1);
 	}
 
@@ -748,17 +773,8 @@
 	 * <!-- end-user-doc -->
 	 * @generated
 	 */
-	public EReference getFeature_Category() {
-		return (EReference)featureEClass.getEStructuralFeatures().get(2);
-	}
-
-	/**
-	 * <!-- begin-user-doc -->
-	 * <!-- end-user-doc -->
-	 * @generated
-	 */
-	public EReference getFeature_Repo() {
-		return (EReference)featureEClass.getEStructuralFeatures().get(3);
+	public EClass getBundle() {
+		return bundleEClass;
 	}
 
 	/**
@@ -766,8 +782,8 @@
 	 * <!-- end-user-doc -->
 	 * @generated
 	 */
-	public EAttribute getFeature_InProduct() {
-		return (EAttribute)featureEClass.getEStructuralFeatures().get(4);
+	public EClass getCompiler() {
+		return compilerEClass;
 	}
 
 	/**
@@ -775,8 +791,8 @@
 	 * <!-- end-user-doc -->
 	 * @generated
 	 */
-	public EClass getBundle() {
-		return bundleEClass;
+	public EAttribute getCompiler_Args() {
+		return (EAttribute)compilerEClass.getEStructuralFeatures().get(0);
 	}
 
 	/**
@@ -784,8 +800,8 @@
 	 * <!-- end-user-doc -->
 	 * @generated
 	 */
-	public EAttribute getBundle_Id() {
-		return (EAttribute)bundleEClass.getEStructuralFeatures().get(0);
+	public EAttribute getCompiler_SourceVersion() {
+		return (EAttribute)compilerEClass.getEStructuralFeatures().get(1);
 	}
 
 	/**
@@ -793,8 +809,8 @@
 	 * <!-- end-user-doc -->
 	 * @generated
 	 */
-	public EAttribute getBundle_Version() {
-		return (EAttribute)bundleEClass.getEStructuralFeatures().get(1);
+	public EAttribute getCompiler_TargetVersion() {
+		return (EAttribute)compilerEClass.getEStructuralFeatures().get(2);
 	}
 
 	/**
@@ -802,8 +818,8 @@
 	 * <!-- end-user-doc -->
 	 * @generated
 	 */
-	public EReference getBundle_Repo() {
-		return (EReference)bundleEClass.getEStructuralFeatures().get(2);
+	public EAttribute getCompiler_Verbose() {
+		return (EAttribute)compilerEClass.getEStructuralFeatures().get(3);
 	}
 
 	/**
@@ -811,8 +827,8 @@
 	 * <!-- end-user-doc -->
 	 * @generated
 	 */
-	public EClass getCompiler() {
-		return compilerEClass;
+	public EAttribute getCompiler_FailOnError() {
+		return (EAttribute)compilerEClass.getEStructuralFeatures().get(4);
 	}
 
 	/**
@@ -820,8 +836,8 @@
 	 * <!-- end-user-doc -->
 	 * @generated
 	 */
-	public EAttribute getCompiler_Args() {
-		return (EAttribute)compilerEClass.getEStructuralFeatures().get(0);
+	public EAttribute getCompiler_DebugInfo() {
+		return (EAttribute)compilerEClass.getEStructuralFeatures().get(5);
 	}
 
 	/**
@@ -829,8 +845,8 @@
 	 * <!-- end-user-doc -->
 	 * @generated
 	 */
-	public EAttribute getCompiler_SourceVersion() {
-		return (EAttribute)compilerEClass.getEStructuralFeatures().get(1);
+	public EClass getPromotion() {
+		return promotionEClass;
 	}
 
 	/**
@@ -838,8 +854,8 @@
 	 * <!-- end-user-doc -->
 	 * @generated
 	 */
-	public EAttribute getCompiler_TargetVersion() {
-		return (EAttribute)compilerEClass.getEStructuralFeatures().get(2);
+	public EAttribute getPromotion_UploadDirectory() {
+		return (EAttribute)promotionEClass.getEStructuralFeatures().get(0);
 	}
 
 	/**
@@ -847,8 +863,8 @@
 	 * <!-- end-user-doc -->
 	 * @generated
 	 */
-	public EAttribute getCompiler_Verbose() {
-		return (EAttribute)compilerEClass.getEStructuralFeatures().get(3);
+	public EAttribute getPromotion_DownloadDirectory() {
+		return (EAttribute)promotionEClass.getEStructuralFeatures().get(1);
 	}
 
 	/**
@@ -856,8 +872,8 @@
 	 * <!-- end-user-doc -->
 	 * @generated
 	 */
-	public EAttribute getCompiler_FailOnError() {
-		return (EAttribute)compilerEClass.getEStructuralFeatures().get(4);
+	public EAttribute getPromotion_Incubating() {
+		return (EAttribute)promotionEClass.getEStructuralFeatures().get(2);
 	}
 
 	/**
@@ -865,8 +881,8 @@
 	 * <!-- end-user-doc -->
 	 * @generated
 	 */
-	public EAttribute getCompiler_DebugInfo() {
-		return (EAttribute)compilerEClass.getEStructuralFeatures().get(5);
+	public EAttribute getPromotion_BaseURL() {
+		return (EAttribute)promotionEClass.getEStructuralFeatures().get(3);
 	}
 
 	/**
@@ -874,8 +890,8 @@
 	 * <!-- end-user-doc -->
 	 * @generated
 	 */
-	public EClass getPromotion() {
-		return promotionEClass;
+	public EAttribute getPromotion_BuildAlias() {
+		return (EAttribute)promotionEClass.getEStructuralFeatures().get(4);
 	}
 
 	/**
@@ -883,8 +899,8 @@
 	 * <!-- end-user-doc -->
 	 * @generated
 	 */
-	public EAttribute getPromotion_UploadDirectory() {
-		return (EAttribute)promotionEClass.getEStructuralFeatures().get(0);
+	public EClass getInstallationUnit() {
+		return installationUnitEClass;
 	}
 
 	/**
@@ -892,8 +908,8 @@
 	 * <!-- end-user-doc -->
 	 * @generated
 	 */
-	public EAttribute getPromotion_DownloadDirectory() {
-		return (EAttribute)promotionEClass.getEStructuralFeatures().get(1);
+	public EAttribute getInstallationUnit_Id() {
+		return (EAttribute)installationUnitEClass.getEStructuralFeatures().get(0);
 	}
 
 	/**
@@ -901,8 +917,8 @@
 	 * <!-- end-user-doc -->
 	 * @generated
 	 */
-	public EAttribute getPromotion_Incubating() {
-		return (EAttribute)promotionEClass.getEStructuralFeatures().get(2);
+	public EAttribute getInstallationUnit_Version() {
+		return (EAttribute)installationUnitEClass.getEStructuralFeatures().get(1);
 	}
 
 	/**
@@ -910,8 +926,8 @@
 	 * <!-- end-user-doc -->
 	 * @generated
 	 */
-	public EAttribute getPromotion_BaseURL() {
-		return (EAttribute)promotionEClass.getEStructuralFeatures().get(3);
+	public EReference getInstallationUnit_Repo() {
+		return (EReference)installationUnitEClass.getEStructuralFeatures().get(2);
 	}
 
 	/**
@@ -919,8 +935,8 @@
 	 * <!-- end-user-doc -->
 	 * @generated
 	 */
-	public EAttribute getPromotion_BuildAlias() {
-		return (EAttribute)promotionEClass.getEStructuralFeatures().get(4);
+	public EClass getProduct() {
+		return productEClass;
 	}
 
 	/**
@@ -1052,22 +1068,17 @@
 		createEAttribute(contributionEClass, CONTRIBUTION__LABEL);
 		createEReference(contributionEClass, CONTRIBUTION__REPOSITORIES);
 		createEReference(contributionEClass, CONTRIBUTION__BUNDLES);
+		createEReference(contributionEClass, CONTRIBUTION__PRODUCTS);
 
 		contactEClass = createEClass(CONTACT);
 		createEAttribute(contactEClass, CONTACT__NAME);
 		createEAttribute(contactEClass, CONTACT__EMAIL);
 
 		featureEClass = createEClass(FEATURE);
-		createEAttribute(featureEClass, FEATURE__ID);
-		createEAttribute(featureEClass, FEATURE__VERSION);
 		createEReference(featureEClass, FEATURE__CATEGORY);
-		createEReference(featureEClass, FEATURE__REPO);
 		createEAttribute(featureEClass, FEATURE__IN_PRODUCT);
 
 		bundleEClass = createEClass(BUNDLE);
-		createEAttribute(bundleEClass, BUNDLE__ID);
-		createEAttribute(bundleEClass, BUNDLE__VERSION);
-		createEReference(bundleEClass, BUNDLE__REPO);
 
 		compilerEClass = createEClass(COMPILER);
 		createEAttribute(compilerEClass, COMPILER__ARGS);
@@ -1084,6 +1095,13 @@
 		createEAttribute(promotionEClass, PROMOTION__BASE_URL);
 		createEAttribute(promotionEClass, PROMOTION__BUILD_ALIAS);
 
+		installationUnitEClass = createEClass(INSTALLATION_UNIT);
+		createEAttribute(installationUnitEClass, INSTALLATION_UNIT__ID);
+		createEAttribute(installationUnitEClass, INSTALLATION_UNIT__VERSION);
+		createEReference(installationUnitEClass, INSTALLATION_UNIT__REPO);
+
+		productEClass = createEClass(PRODUCT);
+
 		// Create enums
 		buildTypeEEnum = createEEnum(BUILD_TYPE);
 		osEEnum = createEEnum(OS);
@@ -1124,6 +1142,9 @@
 		// Set bounds for type parameters
 
 		// Add supertypes to classes
+		featureEClass.getESuperTypes().add(this.getInstallationUnit());
+		bundleEClass.getESuperTypes().add(this.getInstallationUnit());
+		productEClass.getESuperTypes().add(this.getInstallationUnit());
 
 		// Initialize classes and features; add operations and parameters
 		initEClass(buildEClass, Build.class, "Build", !IS_ABSTRACT, !IS_INTERFACE, IS_GENERATED_INSTANCE_CLASS);
@@ -1182,22 +1203,17 @@
 		initEAttribute(getContribution_Label(), ecorePackage.getEString(), "label", null, 0, 1, Contribution.class, !IS_TRANSIENT, !IS_VOLATILE, IS_CHANGEABLE, !IS_UNSETTABLE, !IS_ID, IS_UNIQUE, !IS_DERIVED, IS_ORDERED);
 		initEReference(getContribution_Repositories(), this.getRepository(), null, "repositories", null, 0, -1, Contribution.class, !IS_TRANSIENT, !IS_VOLATILE, IS_CHANGEABLE, IS_COMPOSITE, IS_RESOLVE_PROXIES, !IS_UNSETTABLE, IS_UNIQUE, !IS_DERIVED, IS_ORDERED);
 		initEReference(getContribution_Bundles(), this.getBundle(), null, "bundles", null, 0, -1, Contribution.class, !IS_TRANSIENT, !IS_VOLATILE, IS_CHANGEABLE, IS_COMPOSITE, IS_RESOLVE_PROXIES, !IS_UNSETTABLE, IS_UNIQUE, !IS_DERIVED, IS_ORDERED);
+		initEReference(getContribution_Products(), this.getProduct(), null, "products", null, 0, -1, Contribution.class, !IS_TRANSIENT, !IS_VOLATILE, IS_CHANGEABLE, IS_COMPOSITE, IS_RESOLVE_PROXIES, !IS_UNSETTABLE, IS_UNIQUE, !IS_DERIVED, IS_ORDERED);
 
 		initEClass(contactEClass, Contact.class, "Contact", !IS_ABSTRACT, !IS_INTERFACE, IS_GENERATED_INSTANCE_CLASS);
 		initEAttribute(getContact_Name(), ecorePackage.getEString(), "name", null, 0, 1, Contact.class, !IS_TRANSIENT, !IS_VOLATILE, IS_CHANGEABLE, !IS_UNSETTABLE, !IS_ID, IS_UNIQUE, !IS_DERIVED, IS_ORDERED);
 		initEAttribute(getContact_Email(), ecorePackage.getEString(), "email", null, 0, 1, Contact.class, !IS_TRANSIENT, !IS_VOLATILE, IS_CHANGEABLE, !IS_UNSETTABLE, !IS_ID, IS_UNIQUE, !IS_DERIVED, IS_ORDERED);
 
 		initEClass(featureEClass, Feature.class, "Feature", !IS_ABSTRACT, !IS_INTERFACE, IS_GENERATED_INSTANCE_CLASS);
-		initEAttribute(getFeature_Id(), ecorePackage.getEString(), "id", null, 1, 1, Feature.class, !IS_TRANSIENT, !IS_VOLATILE, IS_CHANGEABLE, !IS_UNSETTABLE, !IS_ID, IS_UNIQUE, !IS_DERIVED, IS_ORDERED);
-		initEAttribute(getFeature_Version(), ecorePackage.getEString(), "version", null, 0, 1, Feature.class, !IS_TRANSIENT, !IS_VOLATILE, IS_CHANGEABLE, !IS_UNSETTABLE, !IS_ID, IS_UNIQUE, !IS_DERIVED, IS_ORDERED);
 		initEReference(getFeature_Category(), this.getCategory(), this.getCategory_Features(), "category", null, 0, -1, Feature.class, !IS_TRANSIENT, !IS_VOLATILE, IS_CHANGEABLE, !IS_COMPOSITE, IS_RESOLVE_PROXIES, !IS_UNSETTABLE, IS_UNIQUE, !IS_DERIVED, IS_ORDERED);
-		initEReference(getFeature_Repo(), this.getRepository(), null, "repo", null, 0, 1, Feature.class, !IS_TRANSIENT, !IS_VOLATILE, IS_CHANGEABLE, !IS_COMPOSITE, IS_RESOLVE_PROXIES, !IS_UNSETTABLE, IS_UNIQUE, !IS_DERIVED, IS_ORDERED);
 		initEAttribute(getFeature_InProduct(), ecorePackage.getEBoolean(), "inProduct", "true", 0, 1, Feature.class, !IS_TRANSIENT, !IS_VOLATILE, IS_CHANGEABLE, !IS_UNSETTABLE, !IS_ID, IS_UNIQUE, !IS_DERIVED, IS_ORDERED);
 
 		initEClass(bundleEClass, Bundle.class, "Bundle", !IS_ABSTRACT, !IS_INTERFACE, IS_GENERATED_INSTANCE_CLASS);
-		initEAttribute(getBundle_Id(), ecorePackage.getEString(), "id", null, 1, 1, Bundle.class, !IS_TRANSIENT, !IS_VOLATILE, IS_CHANGEABLE, !IS_UNSETTABLE, !IS_ID, IS_UNIQUE, !IS_DERIVED, IS_ORDERED);
-		initEAttribute(getBundle_Version(), ecorePackage.getEString(), "version", null, 0, 1, Bundle.class, !IS_TRANSIENT, !IS_VOLATILE, IS_CHANGEABLE, !IS_UNSETTABLE, !IS_ID, IS_UNIQUE, !IS_DERIVED, IS_ORDERED);
-		initEReference(getBundle_Repo(), this.getRepository(), null, "repo", null, 0, 1, Bundle.class, !IS_TRANSIENT, !IS_VOLATILE, IS_CHANGEABLE, !IS_COMPOSITE, IS_RESOLVE_PROXIES, !IS_UNSETTABLE, IS_UNIQUE, !IS_DERIVED, IS_ORDERED);
 
 		initEClass(compilerEClass, org.eclipse.amalgam.releng.build.Compiler.class, "Compiler", !IS_ABSTRACT, !IS_INTERFACE, IS_GENERATED_INSTANCE_CLASS);
 		initEAttribute(getCompiler_Args(), ecorePackage.getEString(), "args", null, 0, 1, org.eclipse.amalgam.releng.build.Compiler.class, !IS_TRANSIENT, !IS_VOLATILE, IS_CHANGEABLE, !IS_UNSETTABLE, !IS_ID, IS_UNIQUE, !IS_DERIVED, IS_ORDERED);
@@ -1214,6 +1230,13 @@
 		initEAttribute(getPromotion_BaseURL(), ecorePackage.getEString(), "baseURL", null, 0, 1, Promotion.class, !IS_TRANSIENT, !IS_VOLATILE, IS_CHANGEABLE, !IS_UNSETTABLE, !IS_ID, IS_UNIQUE, !IS_DERIVED, IS_ORDERED);
 		initEAttribute(getPromotion_BuildAlias(), ecorePackage.getEString(), "buildAlias", null, 0, 1, Promotion.class, !IS_TRANSIENT, !IS_VOLATILE, IS_CHANGEABLE, !IS_UNSETTABLE, !IS_ID, IS_UNIQUE, !IS_DERIVED, IS_ORDERED);
 
+		initEClass(installationUnitEClass, InstallationUnit.class, "InstallationUnit", IS_ABSTRACT, !IS_INTERFACE, IS_GENERATED_INSTANCE_CLASS);
+		initEAttribute(getInstallationUnit_Id(), ecorePackage.getEString(), "id", null, 1, 1, InstallationUnit.class, !IS_TRANSIENT, !IS_VOLATILE, IS_CHANGEABLE, !IS_UNSETTABLE, !IS_ID, IS_UNIQUE, !IS_DERIVED, IS_ORDERED);
+		initEAttribute(getInstallationUnit_Version(), ecorePackage.getEString(), "version", null, 0, 1, InstallationUnit.class, !IS_TRANSIENT, !IS_VOLATILE, IS_CHANGEABLE, !IS_UNSETTABLE, !IS_ID, IS_UNIQUE, !IS_DERIVED, IS_ORDERED);
+		initEReference(getInstallationUnit_Repo(), this.getRepository(), null, "repo", null, 0, 1, InstallationUnit.class, !IS_TRANSIENT, !IS_VOLATILE, IS_CHANGEABLE, !IS_COMPOSITE, IS_RESOLVE_PROXIES, !IS_UNSETTABLE, IS_UNIQUE, !IS_DERIVED, IS_ORDERED);
+
+		initEClass(productEClass, Product.class, "Product", !IS_ABSTRACT, !IS_INTERFACE, IS_GENERATED_INSTANCE_CLASS);
+
 		// Initialize enums and add enum literals
 		initEEnum(buildTypeEEnum, BuildType.class, "BuildType");
 		addEEnumLiteral(buildTypeEEnum, BuildType.CONTINUOUS);
Index: src/org/eclipse/amalgam/releng/build/impl/BundleImpl.java
===================================================================
RCS file: /cvsroot/modeling/org.eclipse.amalgam/releng/org.eclipse.amalgam.releng.build.model/src/org/eclipse/amalgam/releng/build/impl/BundleImpl.java,v
retrieving revision 1.1
diff -u -r1.1 BundleImpl.java
--- src/org/eclipse/amalgam/releng/build/impl/BundleImpl.java	24 Nov 2008 20:36:44 -0000	1.1
+++ src/org/eclipse/amalgam/releng/build/impl/BundleImpl.java	20 May 2009 06:08:58 -0000
@@ -30,67 +30,11 @@
  * An implementation of the model object '<em><b>Bundle</b></em>'.
  * <!-- end-user-doc -->
  * <p>
- * The following features are implemented:
- * <ul>
- *   <li>{@link org.eclipse.amalgam.releng.build.impl.BundleImpl#getId <em>Id</em>}</li>
- *   <li>{@link org.eclipse.amalgam.releng.build.impl.BundleImpl#getVersion <em>Version</em>}</li>
- *   <li>{@link org.eclipse.amalgam.releng.build.impl.BundleImpl#getRepo <em>Repo</em>}</li>
- * </ul>
  * </p>
  *
  * @generated
  */
-public class BundleImpl extends EObjectImpl implements Bundle {
-	/**
-	 * The default value of the '{@link #getId() <em>Id</em>}' attribute.
-	 * <!-- begin-user-doc -->
-	 * <!-- end-user-doc -->
-	 * @see #getId()
-	 * @generated
-	 * @ordered
-	 */
-	protected static final String ID_EDEFAULT = null;
-
-	/**
-	 * The cached value of the '{@link #getId() <em>Id</em>}' attribute.
-	 * <!-- begin-user-doc -->
-	 * <!-- end-user-doc -->
-	 * @see #getId()
-	 * @generated
-	 * @ordered
-	 */
-	protected String id = ID_EDEFAULT;
-
-	/**
-	 * The default value of the '{@link #getVersion() <em>Version</em>}' attribute.
-	 * <!-- begin-user-doc -->
-	 * <!-- end-user-doc -->
-	 * @see #getVersion()
-	 * @generated
-	 * @ordered
-	 */
-	protected static final String VERSION_EDEFAULT = null;
-
-	/**
-	 * The cached value of the '{@link #getVersion() <em>Version</em>}' attribute.
-	 * <!-- begin-user-doc -->
-	 * <!-- end-user-doc -->
-	 * @see #getVersion()
-	 * @generated
-	 * @ordered
-	 */
-	protected String version = VERSION_EDEFAULT;
-
-	/**
-	 * The cached value of the '{@link #getRepo() <em>Repo</em>}' reference.
-	 * <!-- begin-user-doc -->
-	 * <!-- end-user-doc -->
-	 * @see #getRepo()
-	 * @generated
-	 * @ordered
-	 */
-	protected Repository repo;
-
+public class BundleImpl extends InstallationUnitImpl implements Bundle {
 	/**
 	 * <!-- begin-user-doc -->
 	 * <!-- end-user-doc -->
@@ -110,181 +54,4 @@
 		return BuildPackage.Literals.BUNDLE;
 	}
 
-	/**
-	 * <!-- begin-user-doc -->
-	 * <!-- end-user-doc -->
-	 * @generated
-	 */
-	public String getId() {
-		return id;
-	}
-
-	/**
-	 * <!-- begin-user-doc -->
-	 * <!-- end-user-doc -->
-	 * @generated
-	 */
-	public void setId(String newId) {
-		String oldId = id;
-		id = newId;
-		if (eNotificationRequired())
-			eNotify(new ENotificationImpl(this, Notification.SET, BuildPackage.BUNDLE__ID, oldId, id));
-	}
-
-	/**
-	 * <!-- begin-user-doc -->
-	 * <!-- end-user-doc -->
-	 * @generated
-	 */
-	public String getVersion() {
-		return version;
-	}
-
-	/**
-	 * <!-- begin-user-doc -->
-	 * <!-- end-user-doc -->
-	 * @generated
-	 */
-	public void setVersion(String newVersion) {
-		String oldVersion = version;
-		version = newVersion;
-		if (eNotificationRequired())
-			eNotify(new ENotificationImpl(this, Notification.SET, BuildPackage.BUNDLE__VERSION, oldVersion, version));
-	}
-
-	/**
-	 * <!-- begin-user-doc -->
-	 * <!-- end-user-doc -->
-	 * @generated
-	 */
-	public Repository getRepo() {
-		if (repo != null && repo.eIsProxy()) {
-			InternalEObject oldRepo = (InternalEObject)repo;
-			repo = (Repository)eResolveProxy(oldRepo);
-			if (repo != oldRepo) {
-				if (eNotificationRequired())
-					eNotify(new ENotificationImpl(this, Notification.RESOLVE, BuildPackage.BUNDLE__REPO, oldRepo, repo));
-			}
-		}
-		return repo;
-	}
-
-	/**
-	 * <!-- begin-user-doc -->
-	 * <!-- end-user-doc -->
-	 * @generated
-	 */
-	public Repository basicGetRepo() {
-		return repo;
-	}
-
-	/**
-	 * <!-- begin-user-doc -->
-	 * <!-- end-user-doc -->
-	 * @generated
-	 */
-	public void setRepo(Repository newRepo) {
-		Repository oldRepo = repo;
-		repo = newRepo;
-		if (eNotificationRequired())
-			eNotify(new ENotificationImpl(this, Notification.SET, BuildPackage.BUNDLE__REPO, oldRepo, repo));
-	}
-
-	/**
-	 * <!-- begin-user-doc -->
-	 * <!-- end-user-doc -->
-	 * @generated
-	 */
-	@Override
-	public Object eGet(int featureID, boolean resolve, boolean coreType) {
-		switch (featureID) {
-			case BuildPackage.BUNDLE__ID:
-				return getId();
-			case BuildPackage.BUNDLE__VERSION:
-				return getVersion();
-			case BuildPackage.BUNDLE__REPO:
-				if (resolve) return getRepo();
-				return basicGetRepo();
-		}
-		return super.eGet(featureID, resolve, coreType);
-	}
-
-	/**
-	 * <!-- begin-user-doc -->
-	 * <!-- end-user-doc -->
-	 * @generated
-	 */
-	@Override
-	public void eSet(int featureID, Object newValue) {
-		switch (featureID) {
-			case BuildPackage.BUNDLE__ID:
-				setId((String)newValue);
-				return;
-			case BuildPackage.BUNDLE__VERSION:
-				setVersion((String)newValue);
-				return;
-			case BuildPackage.BUNDLE__REPO:
-				setRepo((Repository)newValue);
-				return;
-		}
-		super.eSet(featureID, newValue);
-	}
-
-	/**
-	 * <!-- begin-user-doc -->
-	 * <!-- end-user-doc -->
-	 * @generated
-	 */
-	@Override
-	public void eUnset(int featureID) {
-		switch (featureID) {
-			case BuildPackage.BUNDLE__ID:
-				setId(ID_EDEFAULT);
-				return;
-			case BuildPackage.BUNDLE__VERSION:
-				setVersion(VERSION_EDEFAULT);
-				return;
-			case BuildPackage.BUNDLE__REPO:
-				setRepo((Repository)null);
-				return;
-		}
-		super.eUnset(featureID);
-	}
-
-	/**
-	 * <!-- begin-user-doc -->
-	 * <!-- end-user-doc -->
-	 * @generated
-	 */
-	@Override
-	public boolean eIsSet(int featureID) {
-		switch (featureID) {
-			case BuildPackage.BUNDLE__ID:
-				return ID_EDEFAULT == null ? id != null : !ID_EDEFAULT.equals(id);
-			case BuildPackage.BUNDLE__VERSION:
-				return VERSION_EDEFAULT == null ? version != null : !VERSION_EDEFAULT.equals(version);
-			case BuildPackage.BUNDLE__REPO:
-				return repo != null;
-		}
-		return super.eIsSet(featureID);
-	}
-
-	/**
-	 * <!-- begin-user-doc -->
-	 * <!-- end-user-doc -->
-	 * @generated
-	 */
-	@Override
-	public String toString() {
-		if (eIsProxy()) return super.toString();
-
-		StringBuffer result = new StringBuffer(super.toString());
-		result.append(" (id: ");
-		result.append(id);
-		result.append(", version: ");
-		result.append(version);
-		result.append(')');
-		return result.toString();
-	}
-
 } //BundleImpl
Index: src/org/eclipse/amalgam/releng/build/util/BuildSwitch.java
===================================================================
RCS file: /cvsroot/modeling/org.eclipse.amalgam/releng/org.eclipse.amalgam.releng.build.model/src/org/eclipse/amalgam/releng/build/util/BuildSwitch.java,v
retrieving revision 1.2
diff -u -r1.2 BuildSwitch.java
--- src/org/eclipse/amalgam/releng/build/util/BuildSwitch.java	6 Dec 2008 03:59:50 -0000	1.2
+++ src/org/eclipse/amalgam/releng/build/util/BuildSwitch.java	20 May 2009 06:08:59 -0000
@@ -23,8 +23,10 @@
 import org.eclipse.amalgam.releng.build.Contact;
 import org.eclipse.amalgam.releng.build.Contribution;
 import org.eclipse.amalgam.releng.build.Feature;
+import org.eclipse.amalgam.releng.build.InstallationUnit;
 import org.eclipse.amalgam.releng.build.Map;
 import org.eclipse.amalgam.releng.build.Platform;
+import org.eclipse.amalgam.releng.build.Product;
 import org.eclipse.amalgam.releng.build.Promotion;
 import org.eclipse.amalgam.releng.build.Repository;
 import org.eclipse.emf.ecore.EClass;
@@ -155,12 +157,14 @@
 			case BuildPackage.FEATURE: {
 				Feature feature = (Feature)theEObject;
 				T result = caseFeature(feature);
+				if (result == null) result = caseInstallationUnit(feature);
 				if (result == null) result = defaultCase(theEObject);
 				return result;
 			}
 			case BuildPackage.BUNDLE: {
 				Bundle bundle = (Bundle)theEObject;
 				T result = caseBundle(bundle);
+				if (result == null) result = caseInstallationUnit(bundle);
 				if (result == null) result = defaultCase(theEObject);
 				return result;
 			}
@@ -176,6 +180,19 @@
 				if (result == null) result = defaultCase(theEObject);
 				return result;
 			}
+			case BuildPackage.INSTALLATION_UNIT: {
+				InstallationUnit installationUnit = (InstallationUnit)theEObject;
+				T result = caseInstallationUnit(installationUnit);
+				if (result == null) result = defaultCase(theEObject);
+				return result;
+			}
+			case BuildPackage.PRODUCT: {
+				Product product = (Product)theEObject;
+				T result = caseProduct(product);
+				if (result == null) result = caseInstallationUnit(product);
+				if (result == null) result = defaultCase(theEObject);
+				return result;
+			}
 			default: return defaultCase(theEObject);
 		}
 	}
@@ -361,6 +378,36 @@
 	}
 
 	/**
+	 * Returns the result of interpreting the object as an instance of '<em>Installation Unit</em>'.
+	 * <!-- begin-user-doc -->
+	 * This implementation returns null;
+	 * returning a non-null result will terminate the switch.
+	 * <!-- end-user-doc -->
+	 * @param object the target of the switch.
+	 * @return the result of interpreting the object as an instance of '<em>Installation Unit</em>'.
+	 * @see #doSwitch(org.eclipse.emf.ecore.EObject) doSwitch(EObject)
+	 * @generated
+	 */
+	public T caseInstallationUnit(InstallationUnit object) {
+		return null;
+	}
+
+	/**
+	 * Returns the result of interpreting the object as an instance of '<em>Product</em>'.
+	 * <!-- begin-user-doc -->
+	 * This implementation returns null;
+	 * returning a non-null result will terminate the switch.
+	 * <!-- end-user-doc -->
+	 * @param object the target of the switch.
+	 * @return the result of interpreting the object as an instance of '<em>Product</em>'.
+	 * @see #doSwitch(org.eclipse.emf.ecore.EObject) doSwitch(EObject)
+	 * @generated
+	 */
+	public T caseProduct(Product object) {
+		return null;
+	}
+
+	/**
 	 * Returns the result of interpreting the object as an instance of '<em>EObject</em>'.
 	 * <!-- begin-user-doc -->
 	 * This implementation returns null;
Index: src/org/eclipse/amalgam/releng/build/util/BuildAdapterFactory.java
===================================================================
RCS file: /cvsroot/modeling/org.eclipse.amalgam/releng/org.eclipse.amalgam.releng.build.model/src/org/eclipse/amalgam/releng/build/util/BuildAdapterFactory.java,v
retrieving revision 1.2
diff -u -r1.2 BuildAdapterFactory.java
--- src/org/eclipse/amalgam/releng/build/util/BuildAdapterFactory.java	6 Dec 2008 03:59:50 -0000	1.2
+++ src/org/eclipse/amalgam/releng/build/util/BuildAdapterFactory.java	20 May 2009 06:08:58 -0000
@@ -21,8 +21,10 @@
 import org.eclipse.amalgam.releng.build.Contact;
 import org.eclipse.amalgam.releng.build.Contribution;
 import org.eclipse.amalgam.releng.build.Feature;
+import org.eclipse.amalgam.releng.build.InstallationUnit;
 import org.eclipse.amalgam.releng.build.Map;
 import org.eclipse.amalgam.releng.build.Platform;
+import org.eclipse.amalgam.releng.build.Product;
 import org.eclipse.amalgam.releng.build.Promotion;
 import org.eclipse.amalgam.releng.build.Repository;
 import org.eclipse.emf.common.notify.Adapter;
@@ -137,6 +139,14 @@
 				return createPromotionAdapter();
 			}
 			@Override
+			public Adapter caseInstallationUnit(InstallationUnit object) {
+				return createInstallationUnitAdapter();
+			}
+			@Override
+			public Adapter caseProduct(Product object) {
+				return createProductAdapter();
+			}
+			@Override
 			public Adapter defaultCase(EObject object) {
 				return createEObjectAdapter();
 			}
@@ -325,6 +335,34 @@
 	}
 
 	/**
+	 * Creates a new adapter for an object of class '{@link org.eclipse.amalgam.releng.build.InstallationUnit <em>Installation Unit</em>}'.
+	 * <!-- begin-user-doc -->
+	 * This default implementation returns null so that we can easily ignore cases;
+	 * it's useful to ignore a case when inheritance will catch all the cases anyway.
+	 * <!-- end-user-doc -->
+	 * @return the new adapter.
+	 * @see org.eclipse.amalgam.releng.build.InstallationUnit
+	 * @generated
+	 */
+	public Adapter createInstallationUnitAdapter() {
+		return null;
+	}
+
+	/**
+	 * Creates a new adapter for an object of class '{@link org.eclipse.amalgam.releng.build.Product <em>Product</em>}'.
+	 * <!-- begin-user-doc -->
+	 * This default implementation returns null so that we can easily ignore cases;
+	 * it's useful to ignore a case when inheritance will catch all the cases anyway.
+	 * <!-- end-user-doc -->
+	 * @return the new adapter.
+	 * @see org.eclipse.amalgam.releng.build.Product
+	 * @generated
+	 */
+	public Adapter createProductAdapter() {
+		return null;
+	}
+
+	/**
 	 * Creates a new adapter for the default case.
 	 * <!-- begin-user-doc -->
 	 * This default implementation returns null.
Index: src/org/eclipse/amalgam/releng/build/Bundle.java
===================================================================
RCS file: /cvsroot/modeling/org.eclipse.amalgam/releng/org.eclipse.amalgam.releng.build.model/src/org/eclipse/amalgam/releng/build/Bundle.java,v
retrieving revision 1.1
diff -u -r1.1 Bundle.java
--- src/org/eclipse/amalgam/releng/build/Bundle.java	24 Nov 2008 20:36:44 -0000	1.1
+++ src/org/eclipse/amalgam/releng/build/Bundle.java	20 May 2009 06:08:58 -0000
@@ -20,96 +20,11 @@
  * A representation of the model object '<em><b>Bundle</b></em>'.
  * <!-- end-user-doc -->
  *
- * <p>
- * The following features are supported:
- * <ul>
- *   <li>{@link org.eclipse.amalgam.releng.build.Bundle#getId <em>Id</em>}</li>
- *   <li>{@link org.eclipse.amalgam.releng.build.Bundle#getVersion <em>Version</em>}</li>
- *   <li>{@link org.eclipse.amalgam.releng.build.Bundle#getRepo <em>Repo</em>}</li>
- * </ul>
- * </p>
  *
  * @see org.eclipse.amalgam.releng.build.BuildPackage#getBundle()
  * @model
  * @generated
  */
-public interface Bundle extends EObject {
-	/**
-	 * Returns the value of the '<em><b>Id</b></em>' attribute.
-	 * <!-- begin-user-doc -->
-	 * <p>
-	 * If the meaning of the '<em>Id</em>' attribute isn't clear,
-	 * there really should be more of a description here...
-	 * </p>
-	 * <!-- end-user-doc -->
-	 * @return the value of the '<em>Id</em>' attribute.
-	 * @see #setId(String)
-	 * @see org.eclipse.amalgam.releng.build.BuildPackage#getBundle_Id()
-	 * @model required="true"
-	 * @generated
-	 */
-	String getId();
-
-	/**
-	 * Sets the value of the '{@link org.eclipse.amalgam.releng.build.Bundle#getId <em>Id</em>}' attribute.
-	 * <!-- begin-user-doc -->
-	 * <!-- end-user-doc -->
-	 * @param value the new value of the '<em>Id</em>' attribute.
-	 * @see #getId()
-	 * @generated
-	 */
-	void setId(String value);
-
-	/**
-	 * Returns the value of the '<em><b>Version</b></em>' attribute.
-	 * <!-- begin-user-doc -->
-	 * <p>
-	 * If the meaning of the '<em>Version</em>' attribute isn't clear,
-	 * there really should be more of a description here...
-	 * </p>
-	 * <!-- end-user-doc -->
-	 * @return the value of the '<em>Version</em>' attribute.
-	 * @see #setVersion(String)
-	 * @see org.eclipse.amalgam.releng.build.BuildPackage#getBundle_Version()
-	 * @model
-	 * @generated
-	 */
-	String getVersion();
-
-	/**
-	 * Sets the value of the '{@link org.eclipse.amalgam.releng.build.Bundle#getVersion <em>Version</em>}' attribute.
-	 * <!-- begin-user-doc -->
-	 * <!-- end-user-doc -->
-	 * @param value the new value of the '<em>Version</em>' attribute.
-	 * @see #getVersion()
-	 * @generated
-	 */
-	void setVersion(String value);
-
-	/**
-	 * Returns the value of the '<em><b>Repo</b></em>' reference.
-	 * <!-- begin-user-doc -->
-	 * <p>
-	 * If the meaning of the '<em>Repo</em>' reference isn't clear,
-	 * there really should be more of a description here...
-	 * </p>
-	 * <!-- end-user-doc -->
-	 * @return the value of the '<em>Repo</em>' reference.
-	 * @see #setRepo(Repository)
-	 * @see org.eclipse.amalgam.releng.build.BuildPackage#getBundle_Repo()
-	 * @model
-	 * @generated
-	 */
-	Repository getRepo();
-
-	/**
-	 * Sets the value of the '{@link org.eclipse.amalgam.releng.build.Bundle#getRepo <em>Repo</em>}' reference.
-	 * <!-- begin-user-doc -->
-	 * <!-- end-user-doc -->
-	 * @param value the new value of the '<em>Repo</em>' reference.
-	 * @see #getRepo()
-	 * @generated
-	 */
-	void setRepo(Repository value);
+public interface Bundle extends InstallationUnit {
 
 } // Bundle
Index: src/org/eclipse/amalgam/releng/build/BuildFactory.java
===================================================================
RCS file: /cvsroot/modeling/org.eclipse.amalgam/releng/org.eclipse.amalgam.releng.build.model/src/org/eclipse/amalgam/releng/build/BuildFactory.java,v
retrieving revision 1.3
diff -u -r1.3 BuildFactory.java
--- src/org/eclipse/amalgam/releng/build/BuildFactory.java	13 Mar 2009 11:09:27 -0000	1.3
+++ src/org/eclipse/amalgam/releng/build/BuildFactory.java	20 May 2009 06:08:56 -0000
@@ -141,6 +141,15 @@
 	Promotion createPromotion();
 
 	/**
+	 * Returns a new object of class '<em>Product</em>'.
+	 * <!-- begin-user-doc -->
+	 * <!-- end-user-doc -->
+	 * @return a new object of class '<em>Product</em>'.
+	 * @generated
+	 */
+	Product createProduct();
+
+	/**
 	 * Returns the package supported by this factory.
 	 * <!-- begin-user-doc -->
 	 * <!-- end-user-doc -->
Index: src/org/eclipse/amalgam/releng/build/Feature.java
===================================================================
RCS file: /cvsroot/modeling/org.eclipse.amalgam/releng/org.eclipse.amalgam.releng.build.model/src/org/eclipse/amalgam/releng/build/Feature.java,v
retrieving revision 1.2
diff -u -r1.2 Feature.java
--- src/org/eclipse/amalgam/releng/build/Feature.java	17 Apr 2009 23:42:56 -0000	1.2
+++ src/org/eclipse/amalgam/releng/build/Feature.java	20 May 2009 06:08:58 -0000
@@ -24,10 +24,7 @@
  * <p>
  * The following features are supported:
  * <ul>
- *   <li>{@link org.eclipse.amalgam.releng.build.Feature#getId <em>Id</em>}</li>
- *   <li>{@link org.eclipse.amalgam.releng.build.Feature#getVersion <em>Version</em>}</li>
  *   <li>{@link org.eclipse.amalgam.releng.build.Feature#getCategory <em>Category</em>}</li>
- *   <li>{@link org.eclipse.amalgam.releng.build.Feature#getRepo <em>Repo</em>}</li>
  *   <li>{@link org.eclipse.amalgam.releng.build.Feature#isInProduct <em>In Product</em>}</li>
  * </ul>
  * </p>
@@ -36,59 +33,7 @@
  * @model
  * @generated
  */
-public interface Feature extends EObject {
-	/**
-	 * Returns the value of the '<em><b>Id</b></em>' attribute.
-	 * <!-- begin-user-doc -->
-	 * <p>
-	 * If the meaning of the '<em>Id</em>' attribute isn't clear,
-	 * there really should be more of a description here...
-	 * </p>
-	 * <!-- end-user-doc -->
-	 * @return the value of the '<em>Id</em>' attribute.
-	 * @see #setId(String)
-	 * @see org.eclipse.amalgam.releng.build.BuildPackage#getFeature_Id()
-	 * @model required="true"
-	 * @generated
-	 */
-	String getId();
-
-	/**
-	 * Sets the value of the '{@link org.eclipse.amalgam.releng.build.Feature#getId <em>Id</em>}' attribute.
-	 * <!-- begin-user-doc -->
-	 * <!-- end-user-doc -->
-	 * @param value the new value of the '<em>Id</em>' attribute.
-	 * @see #getId()
-	 * @generated
-	 */
-	void setId(String value);
-
-	/**
-	 * Returns the value of the '<em><b>Version</b></em>' attribute.
-	 * <!-- begin-user-doc -->
-	 * <p>
-	 * If the meaning of the '<em>Version</em>' attribute isn't clear,
-	 * there really should be more of a description here...
-	 * </p>
-	 * <!-- end-user-doc -->
-	 * @return the value of the '<em>Version</em>' attribute.
-	 * @see #setVersion(String)
-	 * @see org.eclipse.amalgam.releng.build.BuildPackage#getFeature_Version()
-	 * @model
-	 * @generated
-	 */
-	String getVersion();
-
-	/**
-	 * Sets the value of the '{@link org.eclipse.amalgam.releng.build.Feature#getVersion <em>Version</em>}' attribute.
-	 * <!-- begin-user-doc -->
-	 * <!-- end-user-doc -->
-	 * @param value the new value of the '<em>Version</em>' attribute.
-	 * @see #getVersion()
-	 * @generated
-	 */
-	void setVersion(String value);
-
+public interface Feature extends InstallationUnit {
 	/**
 	 * Returns the value of the '<em><b>Category</b></em>' reference list.
 	 * The list contents are of type {@link org.eclipse.amalgam.releng.build.Category}.
@@ -108,32 +53,6 @@
 	EList<Category> getCategory();
 
 	/**
-	 * Returns the value of the '<em><b>Repo</b></em>' reference.
-	 * <!-- begin-user-doc -->
-	 * <p>
-	 * If the meaning of the '<em>Repo</em>' reference isn't clear,
-	 * there really should be more of a description here...
-	 * </p>
-	 * <!-- end-user-doc -->
-	 * @return the value of the '<em>Repo</em>' reference.
-	 * @see #setRepo(Repository)
-	 * @see org.eclipse.amalgam.releng.build.BuildPackage#getFeature_Repo()
-	 * @model
-	 * @generated
-	 */
-	Repository getRepo();
-
-	/**
-	 * Sets the value of the '{@link org.eclipse.amalgam.releng.build.Feature#getRepo <em>Repo</em>}' reference.
-	 * <!-- begin-user-doc -->
-	 * <!-- end-user-doc -->
-	 * @param value the new value of the '<em>Repo</em>' reference.
-	 * @see #getRepo()
-	 * @generated
-	 */
-	void setRepo(Repository value);
-
-	/**
 	 * Returns the value of the '<em><b>In Product</b></em>' attribute.
 	 * The default value is <code>"true"</code>.
 	 * <!-- begin-user-doc -->
Index: src/org/eclipse/amalgam/releng/build/Contribution.java
===================================================================
RCS file: /cvsroot/modeling/org.eclipse.amalgam/releng/org.eclipse.amalgam.releng.build.model/src/org/eclipse/amalgam/releng/build/Contribution.java,v
retrieving revision 1.1
diff -u -r1.1 Contribution.java
--- src/org/eclipse/amalgam/releng/build/Contribution.java	24 Nov 2008 20:36:44 -0000	1.1
+++ src/org/eclipse/amalgam/releng/build/Contribution.java	20 May 2009 06:08:58 -0000
@@ -30,6 +30,7 @@
  *   <li>{@link org.eclipse.amalgam.releng.build.Contribution#getLabel <em>Label</em>}</li>
  *   <li>{@link org.eclipse.amalgam.releng.build.Contribution#getRepositories <em>Repositories</em>}</li>
  *   <li>{@link org.eclipse.amalgam.releng.build.Contribution#getBundles <em>Bundles</em>}</li>
+ *   <li>{@link org.eclipse.amalgam.releng.build.Contribution#getProducts <em>Products</em>}</li>
  * </ul>
  * </p>
  *
@@ -128,4 +129,20 @@
 	 */
 	EList<Bundle> getBundles();
 
+	/**
+	 * Returns the value of the '<em><b>Products</b></em>' containment reference list.
+	 * The list contents are of type {@link org.eclipse.amalgam.releng.build.Product}.
+	 * <!-- begin-user-doc -->
+	 * <p>
+	 * If the meaning of the '<em>Products</em>' containment reference list isn't clear,
+	 * there really should be more of a description here...
+	 * </p>
+	 * <!-- end-user-doc -->
+	 * @return the value of the '<em>Products</em>' containment reference list.
+	 * @see org.eclipse.amalgam.releng.build.BuildPackage#getContribution_Products()
+	 * @model containment="true" resolveProxies="true"
+	 * @generated
+	 */
+	EList<Product> getProducts();
+
 } // Contribution
Index: src/org/eclipse/amalgam/releng/build/BuildPackage.java
===================================================================
RCS file: /cvsroot/modeling/org.eclipse.amalgam/releng/org.eclipse.amalgam.releng.build.model/src/org/eclipse/amalgam/releng/build/BuildPackage.java,v
retrieving revision 1.9
diff -u -r1.9 BuildPackage.java
--- src/org/eclipse/amalgam/releng/build/BuildPackage.java	17 Apr 2009 23:42:56 -0000	1.9
+++ src/org/eclipse/amalgam/releng/build/BuildPackage.java	20 May 2009 06:08:58 -0000
@@ -579,13 +579,22 @@
 	int CONTRIBUTION__BUNDLES = 4;
 
 	/**
+	 * The feature id for the '<em><b>Products</b></em>' containment reference list.
+	 * <!-- begin-user-doc -->
+	 * <!-- end-user-doc -->
+	 * @generated
+	 * @ordered
+	 */
+	int CONTRIBUTION__PRODUCTS = 5;
+
+	/**
 	 * The number of structural features of the '<em>Contribution</em>' class.
 	 * <!-- begin-user-doc -->
 	 * <!-- end-user-doc -->
 	 * @generated
 	 * @ordered
 	 */
-	int CONTRIBUTION_FEATURE_COUNT = 5;
+	int CONTRIBUTION_FEATURE_COUNT = 6;
 
 	/**
 	 * The meta object id for the '{@link org.eclipse.amalgam.releng.build.impl.ContactImpl <em>Contact</em>}' class.
@@ -625,6 +634,52 @@
 	int CONTACT_FEATURE_COUNT = 2;
 
 	/**
+	 * The meta object id for the '{@link org.eclipse.amalgam.releng.build.impl.InstallationUnitImpl <em>Installation Unit</em>}' class.
+	 * <!-- begin-user-doc -->
+	 * <!-- end-user-doc -->
+	 * @see org.eclipse.amalgam.releng.build.impl.InstallationUnitImpl
+	 * @see org.eclipse.amalgam.releng.build.impl.BuildPackageImpl#getInstallationUnit()
+	 * @generated
+	 */
+	int INSTALLATION_UNIT = 12;
+
+	/**
+	 * The feature id for the '<em><b>Id</b></em>' attribute.
+	 * <!-- begin-user-doc -->
+	 * <!-- end-user-doc -->
+	 * @generated
+	 * @ordered
+	 */
+	int INSTALLATION_UNIT__ID = 0;
+
+	/**
+	 * The feature id for the '<em><b>Version</b></em>' attribute.
+	 * <!-- begin-user-doc -->
+	 * <!-- end-user-doc -->
+	 * @generated
+	 * @ordered
+	 */
+	int INSTALLATION_UNIT__VERSION = 1;
+
+	/**
+	 * The feature id for the '<em><b>Repo</b></em>' reference.
+	 * <!-- begin-user-doc -->
+	 * <!-- end-user-doc -->
+	 * @generated
+	 * @ordered
+	 */
+	int INSTALLATION_UNIT__REPO = 2;
+
+	/**
+	 * The number of structural features of the '<em>Installation Unit</em>' class.
+	 * <!-- begin-user-doc -->
+	 * <!-- end-user-doc -->
+	 * @generated
+	 * @ordered
+	 */
+	int INSTALLATION_UNIT_FEATURE_COUNT = 3;
+
+	/**
 	 * The meta object id for the '{@link org.eclipse.amalgam.releng.build.impl.FeatureImpl <em>Feature</em>}' class.
 	 * <!-- begin-user-doc -->
 	 * <!-- end-user-doc -->
@@ -641,7 +696,7 @@
 	 * @generated
 	 * @ordered
 	 */
-	int FEATURE__ID = 0;
+	int FEATURE__ID = INSTALLATION_UNIT__ID;
 
 	/**
 	 * The feature id for the '<em><b>Version</b></em>' attribute.
@@ -650,25 +705,25 @@
 	 * @generated
 	 * @ordered
 	 */
-	int FEATURE__VERSION = 1;
+	int FEATURE__VERSION = INSTALLATION_UNIT__VERSION;
 
 	/**
-	 * The feature id for the '<em><b>Category</b></em>' reference list.
+	 * The feature id for the '<em><b>Repo</b></em>' reference.
 	 * <!-- begin-user-doc -->
 	 * <!-- end-user-doc -->
 	 * @generated
 	 * @ordered
 	 */
-	int FEATURE__CATEGORY = 2;
+	int FEATURE__REPO = INSTALLATION_UNIT__REPO;
 
 	/**
-	 * The feature id for the '<em><b>Repo</b></em>' reference.
+	 * The feature id for the '<em><b>Category</b></em>' reference list.
 	 * <!-- begin-user-doc -->
 	 * <!-- end-user-doc -->
 	 * @generated
 	 * @ordered
 	 */
-	int FEATURE__REPO = 3;
+	int FEATURE__CATEGORY = INSTALLATION_UNIT_FEATURE_COUNT + 0;
 
 	/**
 	 * The feature id for the '<em><b>In Product</b></em>' attribute.
@@ -677,7 +732,7 @@
 	 * @generated
 	 * @ordered
 	 */
-	int FEATURE__IN_PRODUCT = 4;
+	int FEATURE__IN_PRODUCT = INSTALLATION_UNIT_FEATURE_COUNT + 1;
 
 	/**
 	 * The number of structural features of the '<em>Feature</em>' class.
@@ -686,7 +741,7 @@
 	 * @generated
 	 * @ordered
 	 */
-	int FEATURE_FEATURE_COUNT = 5;
+	int FEATURE_FEATURE_COUNT = INSTALLATION_UNIT_FEATURE_COUNT + 2;
 
 	/**
 	 * The meta object id for the '{@link org.eclipse.amalgam.releng.build.impl.BundleImpl <em>Bundle</em>}' class.
@@ -705,7 +760,7 @@
 	 * @generated
 	 * @ordered
 	 */
-	int BUNDLE__ID = 0;
+	int BUNDLE__ID = INSTALLATION_UNIT__ID;
 
 	/**
 	 * The feature id for the '<em><b>Version</b></em>' attribute.
@@ -714,7 +769,7 @@
 	 * @generated
 	 * @ordered
 	 */
-	int BUNDLE__VERSION = 1;
+	int BUNDLE__VERSION = INSTALLATION_UNIT__VERSION;
 
 	/**
 	 * The feature id for the '<em><b>Repo</b></em>' reference.
@@ -723,7 +778,7 @@
 	 * @generated
 	 * @ordered
 	 */
-	int BUNDLE__REPO = 2;
+	int BUNDLE__REPO = INSTALLATION_UNIT__REPO;
 
 	/**
 	 * The number of structural features of the '<em>Bundle</em>' class.
@@ -732,7 +787,7 @@
 	 * @generated
 	 * @ordered
 	 */
-	int BUNDLE_FEATURE_COUNT = 3;
+	int BUNDLE_FEATURE_COUNT = INSTALLATION_UNIT_FEATURE_COUNT + 0;
 
 	/**
 	 * The meta object id for the '{@link org.eclipse.amalgam.releng.build.impl.CompilerImpl <em>Compiler</em>}' class.
@@ -872,6 +927,52 @@
 	int PROMOTION_FEATURE_COUNT = 5;
 
 	/**
+	 * The meta object id for the '{@link org.eclipse.amalgam.releng.build.impl.ProductImpl <em>Product</em>}' class.
+	 * <!-- begin-user-doc -->
+	 * <!-- end-user-doc -->
+	 * @see org.eclipse.amalgam.releng.build.impl.ProductImpl
+	 * @see org.eclipse.amalgam.releng.build.impl.BuildPackageImpl#getProduct()
+	 * @generated
+	 */
+	int PRODUCT = 13;
+
+	/**
+	 * The feature id for the '<em><b>Id</b></em>' attribute.
+	 * <!-- begin-user-doc -->
+	 * <!-- end-user-doc -->
+	 * @generated
+	 * @ordered
+	 */
+	int PRODUCT__ID = INSTALLATION_UNIT__ID;
+
+	/**
+	 * The feature id for the '<em><b>Version</b></em>' attribute.
+	 * <!-- begin-user-doc -->
+	 * <!-- end-user-doc -->
+	 * @generated
+	 * @ordered
+	 */
+	int PRODUCT__VERSION = INSTALLATION_UNIT__VERSION;
+
+	/**
+	 * The feature id for the '<em><b>Repo</b></em>' reference.
+	 * <!-- begin-user-doc -->
+	 * <!-- end-user-doc -->
+	 * @generated
+	 * @ordered
+	 */
+	int PRODUCT__REPO = INSTALLATION_UNIT__REPO;
+
+	/**
+	 * The number of structural features of the '<em>Product</em>' class.
+	 * <!-- begin-user-doc -->
+	 * <!-- end-user-doc -->
+	 * @generated
+	 * @ordered
+	 */
+	int PRODUCT_FEATURE_COUNT = INSTALLATION_UNIT_FEATURE_COUNT + 0;
+
+	/**
 	 * The meta object id for the '{@link org.eclipse.amalgam.releng.build.BuildType <em>Type</em>}' enum.
 	 * <!-- begin-user-doc -->
 	 * <!-- end-user-doc -->
@@ -879,7 +980,7 @@
 	 * @see org.eclipse.amalgam.releng.build.impl.BuildPackageImpl#getBuildType()
 	 * @generated
 	 */
-	int BUILD_TYPE = 12;
+	int BUILD_TYPE = 14;
 
 	/**
 	 * The meta object id for the '{@link org.eclipse.amalgam.releng.build.OS <em>OS</em>}' enum.
@@ -889,7 +990,7 @@
 	 * @see org.eclipse.amalgam.releng.build.impl.BuildPackageImpl#getOS()
 	 * @generated
 	 */
-	int OS = 13;
+	int OS = 15;
 
 	/**
 	 * The meta object id for the '{@link org.eclipse.amalgam.releng.build.WS <em>WS</em>}' enum.
@@ -899,7 +1000,7 @@
 	 * @see org.eclipse.amalgam.releng.build.impl.BuildPackageImpl#getWS()
 	 * @generated
 	 */
-	int WS = 14;
+	int WS = 16;
 
 	/**
 	 * The meta object id for the '{@link org.eclipse.amalgam.releng.build.ARCH <em>ARCH</em>}' enum.
@@ -909,7 +1010,7 @@
 	 * @see org.eclipse.amalgam.releng.build.impl.BuildPackageImpl#getARCH()
 	 * @generated
 	 */
-	int ARCH = 15;
+	int ARCH = 17;
 
 	/**
 	 * The meta object id for the '{@link org.eclipse.amalgam.releng.build.ArchiveFormat <em>Archive Format</em>}' enum.
@@ -919,7 +1020,7 @@
 	 * @see org.eclipse.amalgam.releng.build.impl.BuildPackageImpl#getArchiveFormat()
 	 * @generated
 	 */
-	int ARCHIVE_FORMAT = 16;
+	int ARCHIVE_FORMAT = 18;
 
 
 	/**
@@ -1466,6 +1567,17 @@
 	EReference getContribution_Bundles();
 
 	/**
+	 * Returns the meta object for the containment reference list '{@link org.eclipse.amalgam.releng.build.Contribution#getProducts <em>Products</em>}'.
+	 * <!-- begin-user-doc -->
+	 * <!-- end-user-doc -->
+	 * @return the meta object for the containment reference list '<em>Products</em>'.
+	 * @see org.eclipse.amalgam.releng.build.Contribution#getProducts()
+	 * @see #getContribution()
+	 * @generated
+	 */
+	EReference getContribution_Products();
+
+	/**
 	 * Returns the meta object for class '{@link org.eclipse.amalgam.releng.build.Contact <em>Contact</em>}'.
 	 * <!-- begin-user-doc -->
 	 * <!-- end-user-doc -->
@@ -1508,28 +1620,6 @@
 	EClass getFeature();
 
 	/**
-	 * Returns the meta object for the attribute '{@link org.eclipse.amalgam.releng.build.Feature#getId <em>Id</em>}'.
-	 * <!-- begin-user-doc -->
-	 * <!-- end-user-doc -->
-	 * @return the meta object for the attribute '<em>Id</em>'.
-	 * @see org.eclipse.amalgam.releng.build.Feature#getId()
-	 * @see #getFeature()
-	 * @generated
-	 */
-	EAttribute getFeature_Id();
-
-	/**
-	 * Returns the meta object for the attribute '{@link org.eclipse.amalgam.releng.build.Feature#getVersion <em>Version</em>}'.
-	 * <!-- begin-user-doc -->
-	 * <!-- end-user-doc -->
-	 * @return the meta object for the attribute '<em>Version</em>'.
-	 * @see org.eclipse.amalgam.releng.build.Feature#getVersion()
-	 * @see #getFeature()
-	 * @generated
-	 */
-	EAttribute getFeature_Version();
-
-	/**
 	 * Returns the meta object for the reference list '{@link org.eclipse.amalgam.releng.build.Feature#getCategory <em>Category</em>}'.
 	 * <!-- begin-user-doc -->
 	 * <!-- end-user-doc -->
@@ -1541,17 +1631,6 @@
 	EReference getFeature_Category();
 
 	/**
-	 * Returns the meta object for the reference '{@link org.eclipse.amalgam.releng.build.Feature#getRepo <em>Repo</em>}'.
-	 * <!-- begin-user-doc -->
-	 * <!-- end-user-doc -->
-	 * @return the meta object for the reference '<em>Repo</em>'.
-	 * @see org.eclipse.amalgam.releng.build.Feature#getRepo()
-	 * @see #getFeature()
-	 * @generated
-	 */
-	EReference getFeature_Repo();
-
-	/**
 	 * Returns the meta object for the attribute '{@link org.eclipse.amalgam.releng.build.Feature#isInProduct <em>In Product</em>}'.
 	 * <!-- begin-user-doc -->
 	 * <!-- end-user-doc -->
@@ -1573,39 +1652,6 @@
 	EClass getBundle();
 
 	/**
-	 * Returns the meta object for the attribute '{@link org.eclipse.amalgam.releng.build.Bundle#getId <em>Id</em>}'.
-	 * <!-- begin-user-doc -->
-	 * <!-- end-user-doc -->
-	 * @return the meta object for the attribute '<em>Id</em>'.
-	 * @see org.eclipse.amalgam.releng.build.Bundle#getId()
-	 * @see #getBundle()
-	 * @generated
-	 */
-	EAttribute getBundle_Id();
-
-	/**
-	 * Returns the meta object for the attribute '{@link org.eclipse.amalgam.releng.build.Bundle#getVersion <em>Version</em>}'.
-	 * <!-- begin-user-doc -->
-	 * <!-- end-user-doc -->
-	 * @return the meta object for the attribute '<em>Version</em>'.
-	 * @see org.eclipse.amalgam.releng.build.Bundle#getVersion()
-	 * @see #getBundle()
-	 * @generated
-	 */
-	EAttribute getBundle_Version();
-
-	/**
-	 * Returns the meta object for the reference '{@link org.eclipse.amalgam.releng.build.Bundle#getRepo <em>Repo</em>}'.
-	 * <!-- begin-user-doc -->
-	 * <!-- end-user-doc -->
-	 * @return the meta object for the reference '<em>Repo</em>'.
-	 * @see org.eclipse.amalgam.releng.build.Bundle#getRepo()
-	 * @see #getBundle()
-	 * @generated
-	 */
-	EReference getBundle_Repo();
-
-	/**
 	 * Returns the meta object for class '{@link org.eclipse.amalgam.releng.build.Compiler <em>Compiler</em>}'.
 	 * <!-- begin-user-doc -->
 	 * <!-- end-user-doc -->
@@ -1747,6 +1793,59 @@
 	EAttribute getPromotion_BuildAlias();
 
 	/**
+	 * Returns the meta object for class '{@link org.eclipse.amalgam.releng.build.InstallationUnit <em>Installation Unit</em>}'.
+	 * <!-- begin-user-doc -->
+	 * <!-- end-user-doc -->
+	 * @return the meta object for class '<em>Installation Unit</em>'.
+	 * @see org.eclipse.amalgam.releng.build.InstallationUnit
+	 * @generated
+	 */
+	EClass getInstallationUnit();
+
+	/**
+	 * Returns the meta object for the attribute '{@link org.eclipse.amalgam.releng.build.InstallationUnit#getId <em>Id</em>}'.
+	 * <!-- begin-user-doc -->
+	 * <!-- end-user-doc -->
+	 * @return the meta object for the attribute '<em>Id</em>'.
+	 * @see org.eclipse.amalgam.releng.build.InstallationUnit#getId()
+	 * @see #getInstallationUnit()
+	 * @generated
+	 */
+	EAttribute getInstallationUnit_Id();
+
+	/**
+	 * Returns the meta object for the attribute '{@link org.eclipse.amalgam.releng.build.InstallationUnit#getVersion <em>Version</em>}'.
+	 * <!-- begin-user-doc -->
+	 * <!-- end-user-doc -->
+	 * @return the meta object for the attribute '<em>Version</em>'.
+	 * @see org.eclipse.amalgam.releng.build.InstallationUnit#getVersion()
+	 * @see #getInstallationUnit()
+	 * @generated
+	 */
+	EAttribute getInstallationUnit_Version();
+
+	/**
+	 * Returns the meta object for the reference '{@link org.eclipse.amalgam.releng.build.InstallationUnit#getRepo <em>Repo</em>}'.
+	 * <!-- begin-user-doc -->
+	 * <!-- end-user-doc -->
+	 * @return the meta object for the reference '<em>Repo</em>'.
+	 * @see org.eclipse.amalgam.releng.build.InstallationUnit#getRepo()
+	 * @see #getInstallationUnit()
+	 * @generated
+	 */
+	EReference getInstallationUnit_Repo();
+
+	/**
+	 * Returns the meta object for class '{@link org.eclipse.amalgam.releng.build.Product <em>Product</em>}'.
+	 * <!-- begin-user-doc -->
+	 * <!-- end-user-doc -->
+	 * @return the meta object for class '<em>Product</em>'.
+	 * @see org.eclipse.amalgam.releng.build.Product
+	 * @generated
+	 */
+	EClass getProduct();
+
+	/**
 	 * Returns the meta object for enum '{@link org.eclipse.amalgam.releng.build.BuildType <em>Type</em>}'.
 	 * <!-- begin-user-doc -->
 	 * <!-- end-user-doc -->
@@ -2233,6 +2332,14 @@
 		EReference CONTRIBUTION__BUNDLES = eINSTANCE.getContribution_Bundles();
 
 		/**
+		 * The meta object literal for the '<em><b>Products</b></em>' containment reference list feature.
+		 * <!-- begin-user-doc -->
+		 * <!-- end-user-doc -->
+		 * @generated
+		 */
+		EReference CONTRIBUTION__PRODUCTS = eINSTANCE.getContribution_Products();
+
+		/**
 		 * The meta object literal for the '{@link org.eclipse.amalgam.releng.build.impl.ContactImpl <em>Contact</em>}' class.
 		 * <!-- begin-user-doc -->
 		 * <!-- end-user-doc -->
@@ -2269,22 +2376,6 @@
 		EClass FEATURE = eINSTANCE.getFeature();
 
 		/**
-		 * The meta object literal for the '<em><b>Id</b></em>' attribute feature.
-		 * <!-- begin-user-doc -->
-		 * <!-- end-user-doc -->
-		 * @generated
-		 */
-		EAttribute FEATURE__ID = eINSTANCE.getFeature_Id();
-
-		/**
-		 * The meta object literal for the '<em><b>Version</b></em>' attribute feature.
-		 * <!-- begin-user-doc -->
-		 * <!-- end-user-doc -->
-		 * @generated
-		 */
-		EAttribute FEATURE__VERSION = eINSTANCE.getFeature_Version();
-
-		/**
 		 * The meta object literal for the '<em><b>Category</b></em>' reference list feature.
 		 * <!-- begin-user-doc -->
 		 * <!-- end-user-doc -->
@@ -2293,14 +2384,6 @@
 		EReference FEATURE__CATEGORY = eINSTANCE.getFeature_Category();
 
 		/**
-		 * The meta object literal for the '<em><b>Repo</b></em>' reference feature.
-		 * <!-- begin-user-doc -->
-		 * <!-- end-user-doc -->
-		 * @generated
-		 */
-		EReference FEATURE__REPO = eINSTANCE.getFeature_Repo();
-
-		/**
 		 * The meta object literal for the '<em><b>In Product</b></em>' attribute feature.
 		 * <!-- begin-user-doc -->
 		 * <!-- end-user-doc -->
@@ -2319,30 +2402,6 @@
 		EClass BUNDLE = eINSTANCE.getBundle();
 
 		/**
-		 * The meta object literal for the '<em><b>Id</b></em>' attribute feature.
-		 * <!-- begin-user-doc -->
-		 * <!-- end-user-doc -->
-		 * @generated
-		 */
-		EAttribute BUNDLE__ID = eINSTANCE.getBundle_Id();
-
-		/**
-		 * The meta object literal for the '<em><b>Version</b></em>' attribute feature.
-		 * <!-- begin-user-doc -->
-		 * <!-- end-user-doc -->
-		 * @generated
-		 */
-		EAttribute BUNDLE__VERSION = eINSTANCE.getBundle_Version();
-
-		/**
-		 * The meta object literal for the '<em><b>Repo</b></em>' reference feature.
-		 * <!-- begin-user-doc -->
-		 * <!-- end-user-doc -->
-		 * @generated
-		 */
-		EReference BUNDLE__REPO = eINSTANCE.getBundle_Repo();
-
-		/**
 		 * The meta object literal for the '{@link org.eclipse.amalgam.releng.build.impl.CompilerImpl <em>Compiler</em>}' class.
 		 * <!-- begin-user-doc -->
 		 * <!-- end-user-doc -->
@@ -2451,6 +2510,50 @@
 		EAttribute PROMOTION__BUILD_ALIAS = eINSTANCE.getPromotion_BuildAlias();
 
 		/**
+		 * The meta object literal for the '{@link org.eclipse.amalgam.releng.build.impl.InstallationUnitImpl <em>Installation Unit</em>}' class.
+		 * <!-- begin-user-doc -->
+		 * <!-- end-user-doc -->
+		 * @see org.eclipse.amalgam.releng.build.impl.InstallationUnitImpl
+		 * @see org.eclipse.amalgam.releng.build.impl.BuildPackageImpl#getInstallationUnit()
+		 * @generated
+		 */
+		EClass INSTALLATION_UNIT = eINSTANCE.getInstallationUnit();
+
+		/**
+		 * The meta object literal for the '<em><b>Id</b></em>' attribute feature.
+		 * <!-- begin-user-doc -->
+		 * <!-- end-user-doc -->
+		 * @generated
+		 */
+		EAttribute INSTALLATION_UNIT__ID = eINSTANCE.getInstallationUnit_Id();
+
+		/**
+		 * The meta object literal for the '<em><b>Version</b></em>' attribute feature.
+		 * <!-- begin-user-doc -->
+		 * <!-- end-user-doc -->
+		 * @generated
+		 */
+		EAttribute INSTALLATION_UNIT__VERSION = eINSTANCE.getInstallationUnit_Version();
+
+		/**
+		 * The meta object literal for the '<em><b>Repo</b></em>' reference feature.
+		 * <!-- begin-user-doc -->
+		 * <!-- end-user-doc -->
+		 * @generated
+		 */
+		EReference INSTALLATION_UNIT__REPO = eINSTANCE.getInstallationUnit_Repo();
+
+		/**
+		 * The meta object literal for the '{@link org.eclipse.amalgam.releng.build.impl.ProductImpl <em>Product</em>}' class.
+		 * <!-- begin-user-doc -->
+		 * <!-- end-user-doc -->
+		 * @see org.eclipse.amalgam.releng.build.impl.ProductImpl
+		 * @see org.eclipse.amalgam.releng.build.impl.BuildPackageImpl#getProduct()
+		 * @generated
+		 */
+		EClass PRODUCT = eINSTANCE.getProduct();
+
+		/**
 		 * The meta object literal for the '{@link org.eclipse.amalgam.releng.build.BuildType <em>Type</em>}' enum.
 		 * <!-- begin-user-doc -->
 		 * <!-- end-user-doc -->
Index: src/org/eclipse/amalgam/releng/build/impl/ProductImpl.java
===================================================================
RCS file: src/org/eclipse/amalgam/releng/build/impl/ProductImpl.java
diff -N src/org/eclipse/amalgam/releng/build/impl/ProductImpl.java
--- /dev/null	1 Jan 1970 00:00:00 -0000
+++ src/org/eclipse/amalgam/releng/build/impl/ProductImpl.java	1 Jan 1970 00:00:00 -0000
@@ -0,0 +1,50 @@
+/**
+ * Copyright (c) 2008 Borland Software Corporation
+ *   
+ *   All rights reserved. This program and the accompanying materials
+ *   are made available under the terms of the Eclipse Public License v1.0
+ *   which accompanies this distribution, and is available at
+ *   http://www.eclipse.org/legal/epl-v10.html
+ *  
+ *   Contributors:
+ *      Borland Software Corporation - initial API and implementation
+ *
+ * $Id$
+ */
+package org.eclipse.amalgam.releng.build.impl;
+
+import org.eclipse.amalgam.releng.build.BuildPackage;
+import org.eclipse.amalgam.releng.build.Product;
+
+import org.eclipse.emf.ecore.EClass;
+
+/**
+ * <!-- begin-user-doc -->
+ * An implementation of the model object '<em><b>Product</b></em>'.
+ * <!-- end-user-doc -->
+ * <p>
+ * </p>
+ *
+ * @generated
+ */
+public class ProductImpl extends InstallationUnitImpl implements Product {
+	/**
+	 * <!-- begin-user-doc -->
+	 * <!-- end-user-doc -->
+	 * @generated
+	 */
+	protected ProductImpl() {
+		super();
+	}
+
+	/**
+	 * <!-- begin-user-doc -->
+	 * <!-- end-user-doc -->
+	 * @generated
+	 */
+	@Override
+	protected EClass eStaticClass() {
+		return BuildPackage.Literals.PRODUCT;
+	}
+
+} //ProductImpl
Index: src/org/eclipse/amalgam/releng/build/impl/InstallationUnitImpl.java
===================================================================
RCS file: src/org/eclipse/amalgam/releng/build/impl/InstallationUnitImpl.java
diff -N src/org/eclipse/amalgam/releng/build/impl/InstallationUnitImpl.java
--- /dev/null	1 Jan 1970 00:00:00 -0000
+++ src/org/eclipse/amalgam/releng/build/impl/InstallationUnitImpl.java	1 Jan 1970 00:00:00 -0000
@@ -0,0 +1,290 @@
+/**
+ * Copyright (c) 2008 Borland Software Corporation
+ *   
+ *   All rights reserved. This program and the accompanying materials
+ *   are made available under the terms of the Eclipse Public License v1.0
+ *   which accompanies this distribution, and is available at
+ *   http://www.eclipse.org/legal/epl-v10.html
+ *  
+ *   Contributors:
+ *      Borland Software Corporation - initial API and implementation
+ *
+ * $Id$
+ */
+package org.eclipse.amalgam.releng.build.impl;
+
+import org.eclipse.amalgam.releng.build.BuildPackage;
+import org.eclipse.amalgam.releng.build.InstallationUnit;
+import org.eclipse.amalgam.releng.build.Repository;
+
+import org.eclipse.emf.common.notify.Notification;
+
+import org.eclipse.emf.ecore.EClass;
+import org.eclipse.emf.ecore.InternalEObject;
+
+import org.eclipse.emf.ecore.impl.ENotificationImpl;
+import org.eclipse.emf.ecore.impl.EObjectImpl;
+
+/**
+ * <!-- begin-user-doc -->
+ * An implementation of the model object '<em><b>Installation Unit</b></em>'.
+ * <!-- end-user-doc -->
+ * <p>
+ * The following features are implemented:
+ * <ul>
+ *   <li>{@link org.eclipse.amalgam.releng.build.impl.InstallationUnitImpl#getId <em>Id</em>}</li>
+ *   <li>{@link org.eclipse.amalgam.releng.build.impl.InstallationUnitImpl#getVersion <em>Version</em>}</li>
+ *   <li>{@link org.eclipse.amalgam.releng.build.impl.InstallationUnitImpl#getRepo <em>Repo</em>}</li>
+ * </ul>
+ * </p>
+ *
+ * @generated
+ */
+public abstract class InstallationUnitImpl extends EObjectImpl implements InstallationUnit {
+	/**
+	 * The default value of the '{@link #getId() <em>Id</em>}' attribute.
+	 * <!-- begin-user-doc -->
+	 * <!-- end-user-doc -->
+	 * @see #getId()
+	 * @generated
+	 * @ordered
+	 */
+	protected static final String ID_EDEFAULT = null;
+
+	/**
+	 * The cached value of the '{@link #getId() <em>Id</em>}' attribute.
+	 * <!-- begin-user-doc -->
+	 * <!-- end-user-doc -->
+	 * @see #getId()
+	 * @generated
+	 * @ordered
+	 */
+	protected String id = ID_EDEFAULT;
+
+	/**
+	 * The default value of the '{@link #getVersion() <em>Version</em>}' attribute.
+	 * <!-- begin-user-doc -->
+	 * <!-- end-user-doc -->
+	 * @see #getVersion()
+	 * @generated
+	 * @ordered
+	 */
+	protected static final String VERSION_EDEFAULT = null;
+
+	/**
+	 * The cached value of the '{@link #getVersion() <em>Version</em>}' attribute.
+	 * <!-- begin-user-doc -->
+	 * <!-- end-user-doc -->
+	 * @see #getVersion()
+	 * @generated
+	 * @ordered
+	 */
+	protected String version = VERSION_EDEFAULT;
+
+	/**
+	 * The cached value of the '{@link #getRepo() <em>Repo</em>}' reference.
+	 * <!-- begin-user-doc -->
+	 * <!-- end-user-doc -->
+	 * @see #getRepo()
+	 * @generated
+	 * @ordered
+	 */
+	protected Repository repo;
+
+	/**
+	 * <!-- begin-user-doc -->
+	 * <!-- end-user-doc -->
+	 * @generated
+	 */
+	protected InstallationUnitImpl() {
+		super();
+	}
+
+	/**
+	 * <!-- begin-user-doc -->
+	 * <!-- end-user-doc -->
+	 * @generated
+	 */
+	@Override
+	protected EClass eStaticClass() {
+		return BuildPackage.Literals.INSTALLATION_UNIT;
+	}
+
+	/**
+	 * <!-- begin-user-doc -->
+	 * <!-- end-user-doc -->
+	 * @generated
+	 */
+	public String getId() {
+		return id;
+	}
+
+	/**
+	 * <!-- begin-user-doc -->
+	 * <!-- end-user-doc -->
+	 * @generated
+	 */
+	public void setId(String newId) {
+		String oldId = id;
+		id = newId;
+		if (eNotificationRequired())
+			eNotify(new ENotificationImpl(this, Notification.SET, BuildPackage.INSTALLATION_UNIT__ID, oldId, id));
+	}
+
+	/**
+	 * <!-- begin-user-doc -->
+	 * <!-- end-user-doc -->
+	 * @generated
+	 */
+	public String getVersion() {
+		return version;
+	}
+
+	/**
+	 * <!-- begin-user-doc -->
+	 * <!-- end-user-doc -->
+	 * @generated
+	 */
+	public void setVersion(String newVersion) {
+		String oldVersion = version;
+		version = newVersion;
+		if (eNotificationRequired())
+			eNotify(new ENotificationImpl(this, Notification.SET, BuildPackage.INSTALLATION_UNIT__VERSION, oldVersion, version));
+	}
+
+	/**
+	 * <!-- begin-user-doc -->
+	 * <!-- end-user-doc -->
+	 * @generated
+	 */
+	public Repository getRepo() {
+		if (repo != null && repo.eIsProxy()) {
+			InternalEObject oldRepo = (InternalEObject)repo;
+			repo = (Repository)eResolveProxy(oldRepo);
+			if (repo != oldRepo) {
+				if (eNotificationRequired())
+					eNotify(new ENotificationImpl(this, Notification.RESOLVE, BuildPackage.INSTALLATION_UNIT__REPO, oldRepo, repo));
+			}
+		}
+		return repo;
+	}
+
+	/**
+	 * <!-- begin-user-doc -->
+	 * <!-- end-user-doc -->
+	 * @generated
+	 */
+	public Repository basicGetRepo() {
+		return repo;
+	}
+
+	/**
+	 * <!-- begin-user-doc -->
+	 * <!-- end-user-doc -->
+	 * @generated
+	 */
+	public void setRepo(Repository newRepo) {
+		Repository oldRepo = repo;
+		repo = newRepo;
+		if (eNotificationRequired())
+			eNotify(new ENotificationImpl(this, Notification.SET, BuildPackage.INSTALLATION_UNIT__REPO, oldRepo, repo));
+	}
+
+	/**
+	 * <!-- begin-user-doc -->
+	 * <!-- end-user-doc -->
+	 * @generated
+	 */
+	@Override
+	public Object eGet(int featureID, boolean resolve, boolean coreType) {
+		switch (featureID) {
+			case BuildPackage.INSTALLATION_UNIT__ID:
+				return getId();
+			case BuildPackage.INSTALLATION_UNIT__VERSION:
+				return getVersion();
+			case BuildPackage.INSTALLATION_UNIT__REPO:
+				if (resolve) return getRepo();
+				return basicGetRepo();
+		}
+		return super.eGet(featureID, resolve, coreType);
+	}
+
+	/**
+	 * <!-- begin-user-doc -->
+	 * <!-- end-user-doc -->
+	 * @generated
+	 */
+	@Override
+	public void eSet(int featureID, Object newValue) {
+		switch (featureID) {
+			case BuildPackage.INSTALLATION_UNIT__ID:
+				setId((String)newValue);
+				return;
+			case BuildPackage.INSTALLATION_UNIT__VERSION:
+				setVersion((String)newValue);
+				return;
+			case BuildPackage.INSTALLATION_UNIT__REPO:
+				setRepo((Repository)newValue);
+				return;
+		}
+		super.eSet(featureID, newValue);
+	}
+
+	/**
+	 * <!-- begin-user-doc -->
+	 * <!-- end-user-doc -->
+	 * @generated
+	 */
+	@Override
+	public void eUnset(int featureID) {
+		switch (featureID) {
+			case BuildPackage.INSTALLATION_UNIT__ID:
+				setId(ID_EDEFAULT);
+				return;
+			case BuildPackage.INSTALLATION_UNIT__VERSION:
+				setVersion(VERSION_EDEFAULT);
+				return;
+			case BuildPackage.INSTALLATION_UNIT__REPO:
+				setRepo((Repository)null);
+				return;
+		}
+		super.eUnset(featureID);
+	}
+
+	/**
+	 * <!-- begin-user-doc -->
+	 * <!-- end-user-doc -->
+	 * @generated
+	 */
+	@Override
+	public boolean eIsSet(int featureID) {
+		switch (featureID) {
+			case BuildPackage.INSTALLATION_UNIT__ID:
+				return ID_EDEFAULT == null ? id != null : !ID_EDEFAULT.equals(id);
+			case BuildPackage.INSTALLATION_UNIT__VERSION:
+				return VERSION_EDEFAULT == null ? version != null : !VERSION_EDEFAULT.equals(version);
+			case BuildPackage.INSTALLATION_UNIT__REPO:
+				return repo != null;
+		}
+		return super.eIsSet(featureID);
+	}
+
+	/**
+	 * <!-- begin-user-doc -->
+	 * <!-- end-user-doc -->
+	 * @generated
+	 */
+	@Override
+	public String toString() {
+		if (eIsProxy()) return super.toString();
+
+		StringBuffer result = new StringBuffer(super.toString());
+		result.append(" (id: ");
+		result.append(id);
+		result.append(", version: ");
+		result.append(version);
+		result.append(')');
+		return result.toString();
+	}
+
+} //InstallationUnitImpl
Index: src/org/eclipse/amalgam/releng/build/Product.java
===================================================================
RCS file: src/org/eclipse/amalgam/releng/build/Product.java
diff -N src/org/eclipse/amalgam/releng/build/Product.java
--- /dev/null	1 Jan 1970 00:00:00 -0000
+++ src/org/eclipse/amalgam/releng/build/Product.java	1 Jan 1970 00:00:00 -0000
@@ -0,0 +1,28 @@
+/**
+ * Copyright (c) 2008 Borland Software Corporation
+ *   
+ *   All rights reserved. This program and the accompanying materials
+ *   are made available under the terms of the Eclipse Public License v1.0
+ *   which accompanies this distribution, and is available at
+ *   http://www.eclipse.org/legal/epl-v10.html
+ *  
+ *   Contributors:
+ *      Borland Software Corporation - initial API and implementation
+ *
+ * $Id$
+ */
+package org.eclipse.amalgam.releng.build;
+
+
+/**
+ * <!-- begin-user-doc -->
+ * A representation of the model object '<em><b>Product</b></em>'.
+ * <!-- end-user-doc -->
+ *
+ *
+ * @see org.eclipse.amalgam.releng.build.BuildPackage#getProduct()
+ * @model
+ * @generated
+ */
+public interface Product extends InstallationUnit {
+} // Product
Index: src/org/eclipse/amalgam/releng/build/InstallationUnit.java
===================================================================
RCS file: src/org/eclipse/amalgam/releng/build/InstallationUnit.java
diff -N src/org/eclipse/amalgam/releng/build/InstallationUnit.java
--- /dev/null	1 Jan 1970 00:00:00 -0000
+++ src/org/eclipse/amalgam/releng/build/InstallationUnit.java	1 Jan 1970 00:00:00 -0000
@@ -0,0 +1,115 @@
+/**
+ * Copyright (c) 2008 Borland Software Corporation
+ *   
+ *   All rights reserved. This program and the accompanying materials
+ *   are made available under the terms of the Eclipse Public License v1.0
+ *   which accompanies this distribution, and is available at
+ *   http://www.eclipse.org/legal/epl-v10.html
+ *  
+ *   Contributors:
+ *      Borland Software Corporation - initial API and implementation
+ *
+ * $Id$
+ */
+package org.eclipse.amalgam.releng.build;
+
+import org.eclipse.emf.ecore.EObject;
+
+/**
+ * <!-- begin-user-doc -->
+ * A representation of the model object '<em><b>Installation Unit</b></em>'.
+ * <!-- end-user-doc -->
+ *
+ * <p>
+ * The following features are supported:
+ * <ul>
+ *   <li>{@link org.eclipse.amalgam.releng.build.InstallationUnit#getId <em>Id</em>}</li>
+ *   <li>{@link org.eclipse.amalgam.releng.build.InstallationUnit#getVersion <em>Version</em>}</li>
+ *   <li>{@link org.eclipse.amalgam.releng.build.InstallationUnit#getRepo <em>Repo</em>}</li>
+ * </ul>
+ * </p>
+ *
+ * @see org.eclipse.amalgam.releng.build.BuildPackage#getInstallationUnit()
+ * @model abstract="true"
+ * @generated
+ */
+public interface InstallationUnit extends EObject {
+	/**
+	 * Returns the value of the '<em><b>Id</b></em>' attribute.
+	 * <!-- begin-user-doc -->
+	 * <p>
+	 * If the meaning of the '<em>Id</em>' attribute isn't clear,
+	 * there really should be more of a description here...
+	 * </p>
+	 * <!-- end-user-doc -->
+	 * @return the value of the '<em>Id</em>' attribute.
+	 * @see #setId(String)
+	 * @see org.eclipse.amalgam.releng.build.BuildPackage#getInstallationUnit_Id()
+	 * @model required="true"
+	 * @generated
+	 */
+	String getId();
+
+	/**
+	 * Sets the value of the '{@link org.eclipse.amalgam.releng.build.InstallationUnit#getId <em>Id</em>}' attribute.
+	 * <!-- begin-user-doc -->
+	 * <!-- end-user-doc -->
+	 * @param value the new value of the '<em>Id</em>' attribute.
+	 * @see #getId()
+	 * @generated
+	 */
+	void setId(String value);
+
+	/**
+	 * Returns the value of the '<em><b>Version</b></em>' attribute.
+	 * <!-- begin-user-doc -->
+	 * <p>
+	 * If the meaning of the '<em>Version</em>' attribute isn't clear,
+	 * there really should be more of a description here...
+	 * </p>
+	 * <!-- end-user-doc -->
+	 * @return the value of the '<em>Version</em>' attribute.
+	 * @see #setVersion(String)
+	 * @see org.eclipse.amalgam.releng.build.BuildPackage#getInstallationUnit_Version()
+	 * @model
+	 * @generated
+	 */
+	String getVersion();
+
+	/**
+	 * Sets the value of the '{@link org.eclipse.amalgam.releng.build.InstallationUnit#getVersion <em>Version</em>}' attribute.
+	 * <!-- begin-user-doc -->
+	 * <!-- end-user-doc -->
+	 * @param value the new value of the '<em>Version</em>' attribute.
+	 * @see #getVersion()
+	 * @generated
+	 */
+	void setVersion(String value);
+
+	/**
+	 * Returns the value of the '<em><b>Repo</b></em>' reference.
+	 * <!-- begin-user-doc -->
+	 * <p>
+	 * If the meaning of the '<em>Repo</em>' reference isn't clear,
+	 * there really should be more of a description here...
+	 * </p>
+	 * <!-- end-user-doc -->
+	 * @return the value of the '<em>Repo</em>' reference.
+	 * @see #setRepo(Repository)
+	 * @see org.eclipse.amalgam.releng.build.BuildPackage#getInstallationUnit_Repo()
+	 * @model
+	 * @generated
+	 */
+	Repository getRepo();
+
+	/**
+	 * Sets the value of the '{@link org.eclipse.amalgam.releng.build.InstallationUnit#getRepo <em>Repo</em>}' reference.
+	 * <!-- begin-user-doc -->
+	 * <!-- end-user-doc -->
+	 * @param value the new value of the '<em>Repo</em>' reference.
+	 * @see #getRepo()
+	 * @generated
+	 */
+	void setRepo(Repository value);
+
+} // InstallationUnit
#P org.eclipse.amalgam.releng.builder
Index: model/build.genmodel
===================================================================
RCS file: /cvsroot/modeling/org.eclipse.amalgam/releng/org.eclipse.amalgam.releng.builder/model/build.genmodel,v
retrieving revision 1.11
diff -u -r1.11 build.genmodel
--- model/build.genmodel	17 Mar 2009 17:55:06 -0000	1.11
+++ model/build.genmodel	20 May 2009 06:09:01 -0000
@@ -95,23 +95,17 @@
       <genFeatures createChild="false" ecoreFeature="ecore:EAttribute build.ecore#//Contribution/label"/>
       <genFeatures property="None" children="true" createChild="true" ecoreFeature="ecore:EReference build.ecore#//Contribution/repositories"/>
       <genFeatures property="None" children="true" createChild="true" ecoreFeature="ecore:EReference build.ecore#//Contribution/bundles"/>
+      <genFeatures property="None" children="true" createChild="true" ecoreFeature="ecore:EReference build.ecore#//Contribution/products"/>
     </genClasses>
     <genClasses ecoreClass="build.ecore#//Contact">
       <genFeatures createChild="false" ecoreFeature="ecore:EAttribute build.ecore#//Contact/name"/>
       <genFeatures createChild="false" ecoreFeature="ecore:EAttribute build.ecore#//Contact/email"/>
     </genClasses>
     <genClasses ecoreClass="build.ecore#//Feature">
-      <genFeatures createChild="false" ecoreFeature="ecore:EAttribute build.ecore#//Feature/id"/>
-      <genFeatures createChild="false" ecoreFeature="ecore:EAttribute build.ecore#//Feature/version"/>
       <genFeatures notify="false" createChild="false" propertySortChoices="true" ecoreFeature="ecore:EReference build.ecore#//Feature/category"/>
-      <genFeatures notify="false" createChild="false" propertySortChoices="true" ecoreFeature="ecore:EReference build.ecore#//Feature/repo"/>
       <genFeatures createChild="false" ecoreFeature="ecore:EAttribute build.ecore#//Feature/inProduct"/>
     </genClasses>
-    <genClasses ecoreClass="build.ecore#//Bundle">
-      <genFeatures createChild="false" ecoreFeature="ecore:EAttribute build.ecore#//Bundle/id"/>
-      <genFeatures createChild="false" ecoreFeature="ecore:EAttribute build.ecore#//Bundle/version"/>
-      <genFeatures notify="false" createChild="false" propertySortChoices="true" ecoreFeature="ecore:EReference build.ecore#//Bundle/repo"/>
-    </genClasses>
+    <genClasses ecoreClass="build.ecore#//Bundle"/>
     <genClasses ecoreClass="build.ecore#//Compiler">
       <genFeatures createChild="false" ecoreFeature="ecore:EAttribute build.ecore#//Compiler/args"/>
       <genFeatures createChild="false" ecoreFeature="ecore:EAttribute build.ecore#//Compiler/sourceVersion"/>
@@ -127,5 +121,11 @@
       <genFeatures createChild="false" ecoreFeature="ecore:EAttribute build.ecore#//Promotion/baseURL"/>
       <genFeatures createChild="false" ecoreFeature="ecore:EAttribute build.ecore#//Promotion/buildAlias"/>
     </genClasses>
+    <genClasses image="false" ecoreClass="build.ecore#//InstallationUnit">
+      <genFeatures createChild="false" ecoreFeature="ecore:EAttribute build.ecore#//InstallationUnit/id"/>
+      <genFeatures createChild="false" ecoreFeature="ecore:EAttribute build.ecore#//InstallationUnit/version"/>
+      <genFeatures notify="false" createChild="false" propertySortChoices="true" ecoreFeature="ecore:EReference build.ecore#//InstallationUnit/repo"/>
+    </genClasses>
+    <genClasses ecoreClass="build.ecore#//Product"/>
   </genPackages>
 </genmodel:GenModel>
Index: model/build.ecore
===================================================================
RCS file: /cvsroot/modeling/org.eclipse.amalgam/releng/org.eclipse.amalgam.releng.builder/model/build.ecore,v
retrieving revision 1.17
diff -u -r1.17 build.ecore
--- model/build.ecore	17 Apr 2009 23:42:55 -0000	1.17
+++ model/build.ecore	20 May 2009 06:09:01 -0000
@@ -84,17 +84,16 @@
         eType="#//Repository" containment="true"/>
     <eStructuralFeatures xsi:type="ecore:EReference" name="bundles" upperBound="-1"
         eType="#//Bundle" containment="true"/>
+    <eStructuralFeatures xsi:type="ecore:EReference" name="products" upperBound="-1"
+        eType="#//Product" containment="true"/>
   </eClassifiers>
   <eClassifiers xsi:type="ecore:EClass" name="Contact">
     <eStructuralFeatures xsi:type="ecore:EAttribute" name="name" eType="ecore:EDataType http://www.eclipse.org/emf/2002/Ecore#//EString"/>
     <eStructuralFeatures xsi:type="ecore:EAttribute" name="email" eType="ecore:EDataType http://www.eclipse.org/emf/2002/Ecore#//EString"/>
   </eClassifiers>
-  <eClassifiers xsi:type="ecore:EClass" name="Feature">
-    <eStructuralFeatures xsi:type="ecore:EAttribute" name="id" lowerBound="1" eType="ecore:EDataType http://www.eclipse.org/emf/2002/Ecore#//EString"/>
-    <eStructuralFeatures xsi:type="ecore:EAttribute" name="version" eType="ecore:EDataType http://www.eclipse.org/emf/2002/Ecore#//EString"/>
+  <eClassifiers xsi:type="ecore:EClass" name="Feature" eSuperTypes="#//InstallationUnit">
     <eStructuralFeatures xsi:type="ecore:EReference" name="category" upperBound="-1"
         eType="#//Category" eOpposite="#//Category/features"/>
-    <eStructuralFeatures xsi:type="ecore:EReference" name="repo" eType="#//Repository"/>
     <eStructuralFeatures xsi:type="ecore:EAttribute" name="inProduct" eType="ecore:EDataType http://www.eclipse.org/emf/2002/Ecore#//EBoolean"
         defaultValueLiteral="true"/>
   </eClassifiers>
@@ -118,11 +117,7 @@
     <eLiterals name="zip"/>
     <eLiterals name="tar"/>
   </eClassifiers>
-  <eClassifiers xsi:type="ecore:EClass" name="Bundle">
-    <eStructuralFeatures xsi:type="ecore:EAttribute" name="id" lowerBound="1" eType="ecore:EDataType http://www.eclipse.org/emf/2002/Ecore#//EString"/>
-    <eStructuralFeatures xsi:type="ecore:EAttribute" name="version" eType="ecore:EDataType http://www.eclipse.org/emf/2002/Ecore#//EString"/>
-    <eStructuralFeatures xsi:type="ecore:EReference" name="repo" eType="#//Repository"/>
-  </eClassifiers>
+  <eClassifiers xsi:type="ecore:EClass" name="Bundle" eSuperTypes="#//InstallationUnit"/>
   <eClassifiers xsi:type="ecore:EClass" name="Compiler">
     <eStructuralFeatures xsi:type="ecore:EAttribute" name="args" eType="ecore:EDataType http://www.eclipse.org/emf/2002/Ecore#//EString"/>
     <eStructuralFeatures xsi:type="ecore:EAttribute" name="sourceVersion" eType="ecore:EDataType http://www.eclipse.org/emf/2002/Ecore#//EString"
@@ -143,4 +138,10 @@
     <eStructuralFeatures xsi:type="ecore:EAttribute" name="baseURL" eType="ecore:EDataType http://www.eclipse.org/emf/2002/Ecore#//EString"/>
     <eStructuralFeatures xsi:type="ecore:EAttribute" name="buildAlias" eType="ecore:EDataType http://www.eclipse.org/emf/2002/Ecore#//EString"/>
   </eClassifiers>
+  <eClassifiers xsi:type="ecore:EClass" name="InstallationUnit" abstract="true">
+    <eStructuralFeatures xsi:type="ecore:EAttribute" name="id" lowerBound="1" eType="ecore:EDataType http://www.eclipse.org/emf/2002/Ecore#//EString"/>
+    <eStructuralFeatures xsi:type="ecore:EAttribute" name="version" eType="ecore:EDataType http://www.eclipse.org/emf/2002/Ecore#//EString"/>
+    <eStructuralFeatures xsi:type="ecore:EReference" name="repo" eType="#//Repository"/>
+  </eClassifiers>
+  <eClassifiers xsi:type="ecore:EClass" name="Product" eSuperTypes="#//InstallationUnit"/>
 </ecore:EPackage>
